<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Bayesian Inference from the Conditional Genetic Stock Identification Model • rubias</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Bayesian Inference from the Conditional Genetic Stock Identification Model">
<meta name="description" content="Implements Bayesian inference for the conditional genetic stock identification model. It allows inference of mixed fisheries and also simulation of mixtures to predict accuracy. A full description of the underlying methods is available in a recently published article in the Canadian Journal of Fisheries and Aquatic Sciences: &lt;doi:10.1139/cjfas-2018-0016&gt;.">
<meta property="og:description" content="Implements Bayesian inference for the conditional genetic stock identification model. It allows inference of mixed fisheries and also simulation of mixtures to predict accuracy. A full description of the underlying methods is available in a recently published article in the Canadian Journal of Fisheries and Aquatic Sciences: &lt;doi:10.1139/cjfas-2018-0016&gt;.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">rubias</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/rubias-fully-bayesian.html">Using the Fully Bayesian Model in rubias</a></li>
    <li><a class="dropdown-item" href="articles/rubias-overview.html">An Overview of rubias Usage</a></li>
    <li><a class="dropdown-item" href="articles/rubias-underlying-data-structures.html">An Explanation of the Underlying Data Structures in rubias</a></li>
    <li><a class="dropdown-item" href="articles/uncertainty-on-total-catch.html">Summarizing Uncertainty on Stock-Specific Total Catch</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/eriqande/rubias/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="rubias--genetic-stock-identification-gsi-in-the-tidyverse">rubias — genetic stock identification (GSI) in the tidyverse<a class="anchor" aria-label="anchor" href="#rubias--genetic-stock-identification-gsi-in-the-tidyverse"></a>
</h1></div>
<p>16 October, 2025</p>
<ul>
<li>
<a href="#input-data">Input Data</a>
<ul>
<li><a href="#how-about-haploid-markers">How about haploid markers?</a></li>
<li><a href="#an-example-reference-data-file">An example reference data file</a></li>
<li><a href="#an-example-mixture-data-file">An example mixture data file</a></li>
<li><a href="#preliminary-good-practice--check-for-duplicate-individuals">Preliminary good practice — check for duplicate individuals</a></li>
<li><a href="#how-about-known-origin-individuals-in-the-mixture">How about known-origin individuals in the mixture?</a></li>
</ul>
</li>
<li>
<a href="#performing-a-genetic-mixture-analysis">Performing a Genetic Mixture Analysis</a>
<ul>
<li><a href="#setting-the-prior-for-the-mixing-proportions">Setting the prior for the mixing proportions</a></li>
<li><a href="#aggregating-collections-into-reporting-units">Aggregating collections into reporting units</a></li>
<li><a href="#creating-posterior-density-curves-from-the-traces">Creating posterior density curves from the traces</a></li>
<li><a href="#computing-credible-intervals-from-the-traces">Computing Credible Intervals from the Traces</a></li>
<li><a href="#assessing-whether-individuals-are-not-from-any-of-the-reference-populations">Assessing whether individuals are not from any of the reference populations</a></li>
<li><a href="#individuals-of-known-origin-in-the-mixture">Individuals of known origin in the mixture</a></li>
<li><a href="#fully-bayesian-model-with-updating-of-allele-freqencies">Fully Bayesian model (with updating of allele freqencies)</a></li>
</ul>
</li>
<li>
<a href="#assessment-of-genetic-references">Assessment of Genetic References</a>
<ul>
<li><a href="#self-assigning-fish-from-the-reference">Self-assigning fish from the reference</a></li>
<li><a href="#simulated-mixtures-using-a-leave-one-out-type-of-approach">Simulated mixtures using a leave-one-out type of approach</a></li>
<li><a href="#specifying-mixture-proportions-in-assess_reference_loo">Specifying mixture proportions in <code>assess_reference_loo()</code></a></li>
<li><a href="#retrieving-the-individual-simulated-fish-posteriors">Retrieving the individual simulated fish posteriors</a></li>
<li><a href="#changing-the-resampling-unit">Changing the resampling unit</a></li>
<li><a href="#sub-specifying-collection-proportions-or-dirichlet-parameters">“sub-specifying” collection proportions or dirichlet parameters</a></li>
<li>
<a href="#multiple-simulation-scenarios-and-100-simulations">Multiple simulation scenarios and “100% Simulations”</a>
<ul>
<li><a href="#do-it-again-with-100-collections">Do it again with 100% collections</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#bootstrap-corrected-reporting-unit-proportions">Bootstrap-Corrected Reporting Unit Proportions</a></li>
<li><a href="#references">References</a></li>
</ul>
<!-- README.md is generated from README.Rmd. Please edit that file --><!-- badges: start -->
<p>This is an R package for performing genetic stock identification (GSI) and associated tasks. Additionally, it includes a method designed to diagnose and correct a bias recently documented in genetic stock identification. The bias occurs when mixture proportion estimates are desired for groups of populations (reporting units) and the number of populations within each reporting unit are uneven.</p>
<p>In order to run C++ implementations of MCMC, rubias requires the package Rcpp (and now, also, RcppParallel for the baseline resampling option), which in turn requires an Rtools installation (if you are on Windows) or XCode (if you are on a Mac). After cloning into the repository with the above dependencies installed, build &amp; reload the package to view further documentation.</p>
<p>The script “/R-main/coalescent_sim” was used to generate coalescent simulations for bias correction validation. This is unnecessary for testing the applicability of our methods to any particular dataset, which should be done using <code><a href="reference/assess_reference_loo.html">assess_reference_loo()</a></code> and <code><a href="reference/assess_pb_bias_correction.html">assess_pb_bias_correction()</a></code>. <code>coalescent_sim()</code> creates simulated populations using the <code>ms</code> coalescent simulation program, available from the Hudson lab at UChicago, and the <code>GSImulator</code> and <code>ms2geno</code> packages, available at <a href="https://github.com/eriqande" class="external-link uri">https://github.com/eriqande</a>, and so requires more dependencies than the rest of the package.</p>
</div>
<div class="section level1">
<h1 id="input-data">Input Data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h1>
<p>The functions for conducting genetic mixture analysis and for doing simulation assessment to predict the accuracy of a set of genetic markers for genetic stock identification require that genetic data be input as a data frame in a specific format:</p>
<ul>
<li>one row per individual</li>
<li>each locus is represented by two adjacent columns, one for each allele (this package is only configured for diploids, at the moment). Allelic types can be expressed as any number or character</li>
<li>missing data at a locus is expressed with NA values for each gene copy at the locus</li>
<li>if one gene copy is missing from a locus in an indivividual, then both gene copies must be missing at the locus.</li>
<li>the name of the locus is taken to be the column name of the <em>first</em> column of each pair of locus columns. The header on the second column is ignored.</li>
<li>the data frame must have four columns of meta data for each individual:
<ul>
<li>
<code>sample_type</code>: a column telling whether the sample is a <code>reference</code> sample or a <code>mixture</code> sample.</li>
<li>
<code>repunit</code>: the reporting unit that an individual/collection belongs to. This is required if sample_type is <code>reference</code>. And if sample_type is <code>mixture</code> then repunit must be <code>NA</code>.<br>
This must be a character vector. Not a factor. The idea of a “reporting unit” is well-known amongst people doing genetic stock identfication of salmon, but might not be familiar elsewhere. Briefly, a reporting unit is a group of populations (which we call “collections”) that are typically closely related genetically, and which will likely be aggregrated in the results of the GSI exercise.</li>
<li>
<code>collection</code>: for reference samples, the name of the population that the individual is from. For mixture samples, this is the name of the particular sample (i.e. stratum or port that is to be treated together in space and time). This must be a character, not a factor.</li>
<li>
<code>indiv</code> a character vector with the ID of the fish. These must be unique.</li>
</ul>
</li>
<li>When we started developing <code>rubias</code>, we intended to allow both the <code>repunit</code> and the <code>collection</code> columns to be either character vectors or factors. Having them as factors might be desirable if, for example, a certain sort order of the collections or repunits was desired. <em>However</em> at some point it became clear to Eric that, given our approach to converting all the data to a C++ data structure of integers, for rapid analyis, we would be exposing ourselves to greater opportunities for bugginess by allowing <code>repunit</code> and <code>collection</code> to be factors. Accordingly, they <strong>must</strong> be character vectors. If they are not, <code>rubias</code> will throw an error. <strong>Note</strong>: if you do have a specific sort order for your collections or repunits, you can always change them into factors after analysis with <code>rubias</code>. Additionally, you can keep extra columns in your original data frame (for example <code>repunit_f</code> or <code>collection_f</code>) in which the repunits or the collections are stored as factors. See, for example the data file <code>alewife</code>. Or you can just keep a character vector that has the sort order you would like, so as to use it when changing things to factors after <code>rubias</code> analysis. (See, for instance, <code>chinook_repunit_levels</code>.)</li>
<li>The file can have any number of other meta data columns; however, <em>they must all occur in the data frame <strong>before</strong> the columns of genetic data</em>.</li>
<li>When you pass a data frame into any of these functions, you have to tell it which column the genetic data starts in, and it is assumed that all the columns after that one contain genetic data.</li>
<li>If you are doing a mixture analysis, the data frame of mixture fish and of the reference fish must have the same column structure, i.e., they must have exactly the same number of columns with exactly the same column names, in the same order and of the same type.</li>
</ul>
<div class="section level2">
<h2 id="how-about-haploid-markers">How about haploid markers?<a class="anchor" aria-label="anchor" href="#how-about-haploid-markers"></a>
</h2>
<p>At the request of the good folks at ADFG, I introduced a few hacks to allow the input to include markers that are haploid (for example mtDNA haplotypes). To denote a marker as haploid you <em>still give it two columns of data</em> in your data frame, but the second column of the haploid marker must be entirely NAs. When <code>rubias</code> is processing the data and it sees this, it assumes that the marker is haploid and it treats it appropriately.</p>
<p>Note that if you have a diploid marker it typically does not make sense to mark one of the gene copies as missing and the other as non-missing. Accordingly, if you have a diploid marker that records just one of the gene copies as missing in any individual, it is going to throw an error. Likewise, if your haploid marker does not have every single individual with and NA at the second gene copy, then it’s also going to throw an error.</p>
</div>
<div class="section level2">
<h2 id="an-example-reference-data-file">An example reference data file<a class="anchor" aria-label="anchor" href="#an-example-reference-data-file"></a>
</h2>
<p>Here are the meta data columns and the first two loci for eight individuals in the <code>chinook</code> reference data set that comes with the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>  <span class="co"># load up the tidyverse library, we will use it later...</span></span>
<span><span class="co">#&gt; ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──</span></span>
<span><span class="co">#&gt; ✔ dplyr     1.1.4     ✔ readr     2.1.5</span></span>
<span><span class="co">#&gt; ✔ forcats   1.0.0     ✔ stringr   1.5.1</span></span>
<span><span class="co">#&gt; ✔ ggplot2   3.5.1     ✔ tibble    3.2.1</span></span>
<span><span class="co">#&gt; ✔ lubridate 1.9.4     ✔ tidyr     1.3.1</span></span>
<span><span class="co">#&gt; ✔ purrr     1.0.4     </span></span>
<span><span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></span>
<span><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></span>
<span><span class="co">#&gt; ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eriqande/rubias" class="external-link">rubias</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chinook</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   sample_type repunit         collection   indiv   Ots_94857.232 Ots_94857.232.1</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;        &lt;chr&gt;           &lt;int&gt;           &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 reference   CentralValleyfa Feather_H_sp Feathe…             2               2</span></span>
<span><span class="co">#&gt; 2 reference   CentralValleyfa Feather_H_sp Feathe…             2               4</span></span>
<span><span class="co">#&gt; 3 reference   CentralValleyfa Feather_H_sp Feathe…             2               4</span></span>
<span><span class="co">#&gt; 4 reference   CentralValleyfa Feather_H_sp Feathe…             2               4</span></span>
<span><span class="co">#&gt; 5 reference   CentralValleyfa Feather_H_sp Feathe…             2               2</span></span>
<span><span class="co">#&gt; 6 reference   CentralValleyfa Feather_H_sp Feathe…             2               4</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: Ots_102213.210 &lt;int&gt;, Ots_102213.210.1 &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="an-example-mixture-data-file">An example mixture data file<a class="anchor" aria-label="anchor" href="#an-example-mixture-data-file"></a>
</h2>
<p>Here is the same for the mixture data frame that goes along with that reference data set:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chinook_mix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   sample_type repunit collection indiv   Ots_94857.232 Ots_94857.232.1</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;           &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 mixture     &lt;NA&gt;    rec2       T124711             4               2</span></span>
<span><span class="co">#&gt; 2 mixture     &lt;NA&gt;    rec2       T124719             4               2</span></span>
<span><span class="co">#&gt; 3 mixture     &lt;NA&gt;    rec2       T124727             4               4</span></span>
<span><span class="co">#&gt; 4 mixture     &lt;NA&gt;    rec1       T124735             4               4</span></span>
<span><span class="co">#&gt; 5 mixture     &lt;NA&gt;    rec1       T124743             2               2</span></span>
<span><span class="co">#&gt; 6 mixture     &lt;NA&gt;    rec1       T124759             4               2</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: Ots_102213.210 &lt;int&gt;, Ots_102213.210.1 &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preliminary-good-practice--check-for-duplicate-individuals">Preliminary good practice — check for duplicate individuals<a class="anchor" aria-label="anchor" href="#preliminary-good-practice--check-for-duplicate-individuals"></a>
</h2>
<p>Sometimes, for a variety of reasons, an individual’s genotype might appear more than once in a data set. <code>rubias</code> has a quick and dirty function to spot pairs of individuals that share a large number of genotypes. Clearly you only want to look at pairs that don’t have a whole lot of missing data, so one parameter is the fraction of loci that are non-missing in either fish. In our experience with Fluidigm assays, if a fish is missing at &gt; 10% of the SNPs, the remaining genotypes are likely to have a fairly high error rate. So, to look for matching samples, let’s require 85% of the genotypes to be non-missing in both members of the pair. The last parameter is the fraction of non-missing loci at which the pair has the same genotype. We will set that to 0.94 first. Here we see it in action:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine chinook and chinook_mix into one big data frame,</span></span>
<span><span class="co"># but drop the California_Coho collection because Coho all</span></span>
<span><span class="co"># have pretty much the same genotype at these loci!</span></span>
<span><span class="va">chinook_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">chinook</span>, <span class="va">chinook_mix</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">collection</span> <span class="op">!=</span> <span class="st">"California_Coho"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># then toss them into a function.  This takes half a minute or so...</span></span>
<span><span class="va">matchy_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/close_matching_samples.html">close_matching_samples</a></span><span class="op">(</span>D <span class="op">=</span> <span class="va">chinook_all</span>, </span>
<span>                                       gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                       min_frac_non_miss <span class="op">=</span> <span class="fl">0.85</span>, </span>
<span>                                       min_frac_matching <span class="op">=</span> <span class="fl">0.94</span></span>
<span>                                       <span class="op">)</span></span>
<span><span class="co">#&gt; Summary Statistics:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 9510 Individuals in Sample</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 91 Loci: AldB1.122, AldoB4.183, OTNAML12_1.SNP1, Ots_100884.287, Ots_101119.381, Ots_101704.143, Ots_102213.210, Ots_102414.395, Ots_102420.494, Ots_102457.132, Ots_102801.308, Ots_102867.609, Ots_103041.52, Ots_104063.132, Ots_104569.86, Ots_105105.613, Ots_105132.200, Ots_105401.325, Ots_105407.117, Ots_106499.70, Ots_106747.239, Ots_107074.284, Ots_107285.93, Ots_107806.821, Ots_108007.208, Ots_108390.329, Ots_108735.302, Ots_109693.392, Ots_110064.383, Ots_110201.363, Ots_110495.380, Ots_110551.64, Ots_111312.435, Ots_111666.408, Ots_111681.657, Ots_112301.43, Ots_112419.131, Ots_112820.284, Ots_112876.371, Ots_113242.216, Ots_113457.40, Ots_117043.255, Ots_117242.136, Ots_117432.409, Ots_118175.479, Ots_118205.61, Ots_118938.325, Ots_122414.56, Ots_123048.521, Ots_123921.111, Ots_124774.477, Ots_127236.62, Ots_128302.57, Ots_128693.461, Ots_128757.61, Ots_129144.472, Ots_129170.683, Ots_129458.451, Ots_130720.99, Ots_131460.584, Ots_131906.141, Ots_94857.232, Ots_96222.525, Ots_96500.180, Ots_97077.179, Ots_99550.204, Ots_ARNT.195, Ots_AsnRS.60, Ots_aspat.196, Ots_CD59.2, Ots_CD63, Ots_EP.529, Ots_GDH.81x, Ots_HSP90B.385, Ots_MHC1, Ots_mybp.85, Ots_myoD.364, Ots_Ots311.101x, Ots_PGK.54, Ots_Prl2, Ots_RFC2.558, Ots_SClkF2R2.135, Ots_SWS1op.182, Ots_TAPBP, Ots_u07.07.161, Ots_u07.49.290, Ots_u4.92, OTSBMP.2.SNP1, OTSTF1.SNP1, S71.336, unk_526</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 39 Reporting Units: CentralValleyfa, CentralValleysp, CentralValleywi, CaliforniaCoast, KlamathR, NCaliforniaSOregonCoast, RogueR, MidOregonCoast, NOregonCoast, WillametteR, DeschutesRfa, LColumbiaRfa, LColumbiaRsp, MidColumbiaRtule, UColumbiaRsufa, MidandUpperColumbiaRsp, SnakeRfa, SnakeRspsu, NPugetSound, WashingtonCoast, SPugetSound, LFraserR, LThompsonR, EVancouverIs, WVancouverIs, MSkeenaR, MidSkeenaR, LSkeenaR, SSEAlaska, NGulfCoastAlsekR, NGulfCoastKarlukR, TakuR, NSEAlaskaChilkatR, NGulfCoastSitukR, CopperR, SusitnaR, LKuskokwimBristolBay, MidYukon</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 68 Collections: Feather_H_sp, Butte_Cr_Sp, Mill_Cr_sp, Deer_Cr_sp, UpperSacramento_R_sp, Feather_H_fa, Butte_Cr_fa, Mill_Cr_fa, Deer_Cr_fa, Mokelumne_R_fa, Battle_Cr, Sacramento_R_lf, Sacramento_H, Eel_R, Russian_R, Klamath_IGH_fa, Trinity_H_sp, Smith_R, Chetco_R, Cole_Rivers_H, Applegate_Cr, Coquille_R, Umpqua_sp, Nestucca_H, Siuslaw_R, Alsea_R, Nehalem_R, Siletz_R, N_Santiam_H, McKenzie_H, L_Deschutes_R, Cowlitz_H_fa, Cowlitz_H_sp, Kalama_H_sp, Spring_Cr_H, Hanford_Reach, PriestRapids_H, Wells_H, Wenatchee_R, CleElum, Lyons_Ferry_H, Rapid_R_H, McCall_H, Kendall_H_sp, Forks_Cr_H, Soos_H, Marblemount_H_sp, QuinaltLake_f, Harris_R, Birkenhead_H, Spius_H, Big_Qual_H, Robertson_H, Morice_R, Kitwanga_R, L_Kalum_R, LPW_Unuk_R, Goat_Cr, Karluk_R, LittleTatsamenie, Tahini_R, Situk_R, Sinona_Ck, Montana_Ck, George_R, Kanektok_R, Togiak_R, Kantishna_R</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3.85% of allelic data identified as missing</span></span>
<span></span>
<span><span class="co"># see that that looks like:</span></span>
<span><span class="va">matchy_pairs</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">num_non_miss</span><span class="op">)</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">num_match</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 33 × 10</span></span>
<span><span class="co">#&gt;    num_non_miss num_match indiv_1            indiv_2   collection_1 collection_2</span></span>
<span><span class="co">#&gt;           &lt;int&gt;     &lt;int&gt; &lt;chr&gt;              &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;       </span></span>
<span><span class="co">#&gt;  1           91        91 T124864            T124866   rec3         rec1        </span></span>
<span><span class="co">#&gt;  2           91        91 T124864            T125335   rec3         rec3        </span></span>
<span><span class="co">#&gt;  3           91        91 T124866            T125335   rec1         rec3        </span></span>
<span><span class="co">#&gt;  4           91        91 T126402            T126403   rec2         rec2        </span></span>
<span><span class="co">#&gt;  5           91        90 Mill_Cr_sp:0060    Mill_Cr_… Mill_Cr_sp   Mill_Cr_sp  </span></span>
<span><span class="co">#&gt;  6           91        90 Cole_Rivers_H:0002 Cole_Riv… Cole_Rivers… Cole_Rivers…</span></span>
<span><span class="co">#&gt;  7           91        90 Cole_Rivers_H:0017 Cole_Riv… Cole_Rivers… Cole_Rivers…</span></span>
<span><span class="co">#&gt;  8           91        90 Umpqua_sp:0009     Umpqua_s… Umpqua_sp    Umpqua_sp   </span></span>
<span><span class="co">#&gt;  9           91        90 Umpqua_sp:0018     Umpqua_s… Umpqua_sp    Umpqua_sp   </span></span>
<span><span class="co">#&gt; 10           91        90 T125044            T125337   rec2         rec1        </span></span>
<span><span class="co">#&gt; # ℹ 23 more rows</span></span>
<span><span class="co">#&gt; # ℹ 4 more variables: sample_type_1 &lt;chr&gt;, repunit_1 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   sample_type_2 &lt;chr&gt;, repunit_2 &lt;chr&gt;</span></span></code></pre></div>
<p>Check that out. This reveals 33 pairs in the data set that are likely duplicate samples.</p>
<p>If we reduce the min_frac_matching, we get more matches, but these are very unlikely to be the same individual, unless genotyping error rates are very high.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># then toss them into a function.  This takes half a minute or so...</span></span>
<span><span class="va">matchy_pairs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/close_matching_samples.html">close_matching_samples</a></span><span class="op">(</span>D <span class="op">=</span> <span class="va">chinook_all</span>, </span>
<span>                                       gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                       min_frac_non_miss <span class="op">=</span> <span class="fl">0.85</span>, </span>
<span>                                       min_frac_matching <span class="op">=</span> <span class="fl">0.85</span></span>
<span>                                       <span class="op">)</span></span>
<span><span class="co">#&gt; Summary Statistics:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 9510 Individuals in Sample</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 91 Loci: AldB1.122, AldoB4.183, OTNAML12_1.SNP1, Ots_100884.287, Ots_101119.381, Ots_101704.143, Ots_102213.210, Ots_102414.395, Ots_102420.494, Ots_102457.132, Ots_102801.308, Ots_102867.609, Ots_103041.52, Ots_104063.132, Ots_104569.86, Ots_105105.613, Ots_105132.200, Ots_105401.325, Ots_105407.117, Ots_106499.70, Ots_106747.239, Ots_107074.284, Ots_107285.93, Ots_107806.821, Ots_108007.208, Ots_108390.329, Ots_108735.302, Ots_109693.392, Ots_110064.383, Ots_110201.363, Ots_110495.380, Ots_110551.64, Ots_111312.435, Ots_111666.408, Ots_111681.657, Ots_112301.43, Ots_112419.131, Ots_112820.284, Ots_112876.371, Ots_113242.216, Ots_113457.40, Ots_117043.255, Ots_117242.136, Ots_117432.409, Ots_118175.479, Ots_118205.61, Ots_118938.325, Ots_122414.56, Ots_123048.521, Ots_123921.111, Ots_124774.477, Ots_127236.62, Ots_128302.57, Ots_128693.461, Ots_128757.61, Ots_129144.472, Ots_129170.683, Ots_129458.451, Ots_130720.99, Ots_131460.584, Ots_131906.141, Ots_94857.232, Ots_96222.525, Ots_96500.180, Ots_97077.179, Ots_99550.204, Ots_ARNT.195, Ots_AsnRS.60, Ots_aspat.196, Ots_CD59.2, Ots_CD63, Ots_EP.529, Ots_GDH.81x, Ots_HSP90B.385, Ots_MHC1, Ots_mybp.85, Ots_myoD.364, Ots_Ots311.101x, Ots_PGK.54, Ots_Prl2, Ots_RFC2.558, Ots_SClkF2R2.135, Ots_SWS1op.182, Ots_TAPBP, Ots_u07.07.161, Ots_u07.49.290, Ots_u4.92, OTSBMP.2.SNP1, OTSTF1.SNP1, S71.336, unk_526</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 39 Reporting Units: CentralValleyfa, CentralValleysp, CentralValleywi, CaliforniaCoast, KlamathR, NCaliforniaSOregonCoast, RogueR, MidOregonCoast, NOregonCoast, WillametteR, DeschutesRfa, LColumbiaRfa, LColumbiaRsp, MidColumbiaRtule, UColumbiaRsufa, MidandUpperColumbiaRsp, SnakeRfa, SnakeRspsu, NPugetSound, WashingtonCoast, SPugetSound, LFraserR, LThompsonR, EVancouverIs, WVancouverIs, MSkeenaR, MidSkeenaR, LSkeenaR, SSEAlaska, NGulfCoastAlsekR, NGulfCoastKarlukR, TakuR, NSEAlaskaChilkatR, NGulfCoastSitukR, CopperR, SusitnaR, LKuskokwimBristolBay, MidYukon</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 68 Collections: Feather_H_sp, Butte_Cr_Sp, Mill_Cr_sp, Deer_Cr_sp, UpperSacramento_R_sp, Feather_H_fa, Butte_Cr_fa, Mill_Cr_fa, Deer_Cr_fa, Mokelumne_R_fa, Battle_Cr, Sacramento_R_lf, Sacramento_H, Eel_R, Russian_R, Klamath_IGH_fa, Trinity_H_sp, Smith_R, Chetco_R, Cole_Rivers_H, Applegate_Cr, Coquille_R, Umpqua_sp, Nestucca_H, Siuslaw_R, Alsea_R, Nehalem_R, Siletz_R, N_Santiam_H, McKenzie_H, L_Deschutes_R, Cowlitz_H_fa, Cowlitz_H_sp, Kalama_H_sp, Spring_Cr_H, Hanford_Reach, PriestRapids_H, Wells_H, Wenatchee_R, CleElum, Lyons_Ferry_H, Rapid_R_H, McCall_H, Kendall_H_sp, Forks_Cr_H, Soos_H, Marblemount_H_sp, QuinaltLake_f, Harris_R, Birkenhead_H, Spius_H, Big_Qual_H, Robertson_H, Morice_R, Kitwanga_R, L_Kalum_R, LPW_Unuk_R, Goat_Cr, Karluk_R, LittleTatsamenie, Tahini_R, Situk_R, Sinona_Ck, Montana_Ck, George_R, Kanektok_R, Togiak_R, Kantishna_R</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3.85% of allelic data identified as missing</span></span>
<span></span>
<span><span class="co"># see that that looks like:</span></span>
<span><span class="va">matchy_pairs2</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">num_non_miss</span><span class="op">)</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">num_match</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 46 × 10</span></span>
<span><span class="co">#&gt;    num_non_miss num_match indiv_1            indiv_2   collection_1 collection_2</span></span>
<span><span class="co">#&gt;           &lt;int&gt;     &lt;int&gt; &lt;chr&gt;              &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;       </span></span>
<span><span class="co">#&gt;  1           91        91 T124864            T124866   rec3         rec1        </span></span>
<span><span class="co">#&gt;  2           91        91 T124864            T125335   rec3         rec3        </span></span>
<span><span class="co">#&gt;  3           91        91 T124866            T125335   rec1         rec3        </span></span>
<span><span class="co">#&gt;  4           91        91 T126402            T126403   rec2         rec2        </span></span>
<span><span class="co">#&gt;  5           91        90 Mill_Cr_sp:0060    Mill_Cr_… Mill_Cr_sp   Mill_Cr_sp  </span></span>
<span><span class="co">#&gt;  6           91        90 Cole_Rivers_H:0002 Cole_Riv… Cole_Rivers… Cole_Rivers…</span></span>
<span><span class="co">#&gt;  7           91        90 Cole_Rivers_H:0017 Cole_Riv… Cole_Rivers… Cole_Rivers…</span></span>
<span><span class="co">#&gt;  8           91        90 Umpqua_sp:0009     Umpqua_s… Umpqua_sp    Umpqua_sp   </span></span>
<span><span class="co">#&gt;  9           91        90 Umpqua_sp:0018     Umpqua_s… Umpqua_sp    Umpqua_sp   </span></span>
<span><span class="co">#&gt; 10           91        90 T125044            T125337   rec2         rec1        </span></span>
<span><span class="co">#&gt; # ℹ 36 more rows</span></span>
<span><span class="co">#&gt; # ℹ 4 more variables: sample_type_1 &lt;chr&gt;, repunit_1 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   sample_type_2 &lt;chr&gt;, repunit_2 &lt;chr&gt;</span></span></code></pre></div>
<p>A more principled approach would be to use the allele frequencies in each collection and take a likelihood based approach, but this is adequate for finding obvious duplicates.</p>
</div>
<div class="section level2">
<h2 id="how-about-known-origin-individuals-in-the-mixture">How about known-origin individuals in the mixture?<a class="anchor" aria-label="anchor" href="#how-about-known-origin-individuals-in-the-mixture"></a>
</h2>
<p>In some cases, you might know (more or less unambiguously) the origin of some fish in a particular mixture sample. For example, if 10% of the individuals in a mixture carried coded wire tags, then you would want to include them in the sample, but make sure that their collections of origin were hard-coded to be what the CWTs said. Another scenario in which this might occur is when the genetic data were used for parentage-based tagging of the individuals in the mixture sample. In that case, some individuals might be placed with very high confidence to parents. Then, they should be included in the mixture as having come from a known collection. The folks at the DFO in Nanaimo, Canada are doing an amazing job with PBT and wondered if rubias could be modified to deal with the latter situation.</p>
<p>We’ve made some small additions to accommodate this. rubias does not do any actual inference of parentage, but if you know the origin of some fish in the mixture, that can be included in the rubias analysis. The way you do this with the function <code><a href="reference/infer_mixture.html">infer_mixture()</a></code> is to include a column called <code>known_collection</code> in both the reference data frame and the mixture data frame. In the reference data frame, <code>known_collection</code> should just be a copy of the <code>collection</code> column. However, in the mixture data frame each entry in <code>known_collection</code> should be the collection that the individual is known to be from (i.e. using parentage inference or a CWT), or, if the individual is not known to be from any collection, it should be NA. Note that the names of the collections in <code>known_collection</code> must match those found in the <code>collection</code> column in the reference data set.</p>
<p>These modifications are not allowed for the parametric bootstrap (PB) method in <code><a href="reference/infer_mixture.html">infer_mixture()</a></code>.</p>
</div>
</div>
<div class="section level1">
<h1 id="performing-a-genetic-mixture-analysis">Performing a Genetic Mixture Analysis<a class="anchor" aria-label="anchor" href="#performing-a-genetic-mixture-analysis"></a>
</h1>
<p>This is done with the <code>infer_mixture</code> function. In the example data <code>chinook_mix</code> our data consist of fish caught in three different fisheries, <code>rec1</code>, <code>rec2</code>, and <code>rec3</code> as denoted in the collection column. Each of those collections is treated as a separate sample, getting its own mixing proportion estimate. This is how it is run with the default options:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                         mixture <span class="op">=</span> <span class="va">chinook_mix</span>, </span>
<span>                         gen_start_col <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Collating data; compiling reference allele frequencies, etc.   time: 0.83 seconds</span></span>
<span><span class="co">#&gt; Computing reference locus specific means and variances for computing mixture z-scores   time: 0.10 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec2 with 772 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.07 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.54 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec1 with 743 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.52 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec3 with 741 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.51 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span></code></pre></div>
<p>The result comes back as a list of four tidy data frames:</p>
<ol style="list-style-type: decimal">
<li>
<code>mixing_proportions</code>: the mixing proportions. The column <code>pi</code> holds the estimated mixing proportion for each collection.</li>
<li>
<code>indiv_posteriors</code>: this holds, for each individual, the posterior means of group membership in each collection. Column <code>PofZ</code> holds those values. Column <code>log_likelihood</code> holds the log of the probability of the individuals genotype given it is from the collection. Also included are <code>n_non_miss_loci</code> and <code>n_miss_loci</code> which are the number of observed loci and the number of missing loci at the individual. A list column <code>missing_loci</code> contains vectors with the indices (and the names) of the loci that are missing in that individual. It also includes a column <code>z_score</code> which can be used to diagnose fish that don’t belong to any samples in the reference data base (see below).</li>
<li>
<code>mix_prop_traces:</code> MCMC traces of the mixing proportions for each collection. You will use these if you want to make density estimates of the posterior distribution of the mixing proportions or if you want to compute credible intervals.</li>
<li>
<code>bootstrapped_proportions</code>: This is NULL in the above example, but if we had chosen <code>method = "PB"</code> then this would be a tibble of bootstrap-corrected reporting unit mixing proportions.</li>
</ol>
<p>These data frames look like this:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">mix_est</span>, <span class="va">head</span><span class="op">)</span></span>
<span><span class="co">#&gt; $mixing_proportions</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 4</span></span>
<span><span class="co">#&gt;   mixture_collection repunit         collection                  pi</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;           &lt;chr&gt;                    &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 rec2               CentralValleyfa Feather_H_sp         0.0819   </span></span>
<span><span class="co">#&gt; 2 rec2               CentralValleysp Butte_Cr_Sp          0.0000443</span></span>
<span><span class="co">#&gt; 3 rec2               CentralValleysp Mill_Cr_sp           0.0000426</span></span>
<span><span class="co">#&gt; 4 rec2               CentralValleysp Deer_Cr_sp           0.0000383</span></span>
<span><span class="co">#&gt; 5 rec2               CentralValleysp UpperSacramento_R_sp 0.000380 </span></span>
<span><span class="co">#&gt; 6 rec2               CentralValleyfa Feather_H_fa         0.147    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $indiv_posteriors</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 10</span></span>
<span><span class="co">#&gt;   mixture_collection indiv   repunit  collection     PofZ log_likelihood z_score</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;         &lt;dbl&gt;          &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 rec2               T124711 Central… Feather_H… 1.86e-28          -137.   -13.1</span></span>
<span><span class="co">#&gt; 2 rec2               T124711 Central… Feather_H… 9.66e-28          -136.   -12.6</span></span>
<span><span class="co">#&gt; 3 rec2               T124711 Central… Butte_Cr_… 1.53e-24          -130.   -10.5</span></span>
<span><span class="co">#&gt; 4 rec2               T124711 Central… Mill_Cr_fa 2.80e-29          -135.   -11.8</span></span>
<span><span class="co">#&gt; 5 rec2               T124711 Central… Deer_Cr_fa 1.82e-28          -134.   -11.6</span></span>
<span><span class="co">#&gt; 6 rec2               T124711 Central… Mokelumne… 1.94e-27          -134.   -12.3</span></span>
<span><span class="co">#&gt; # ℹ 3 more variables: n_non_miss_loci &lt;int&gt;, n_miss_loci &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   missing_loci &lt;list&gt;</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $mix_prop_traces</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 5</span></span>
<span><span class="co">#&gt;   mixture_collection sweep repunit         collection               pi</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt; &lt;chr&gt;           &lt;chr&gt;                 &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 rec2                   0 CentralValleyfa Feather_H_sp         0.0145</span></span>
<span><span class="co">#&gt; 2 rec2                   0 CentralValleysp Butte_Cr_Sp          0.0145</span></span>
<span><span class="co">#&gt; 3 rec2                   0 CentralValleysp Mill_Cr_sp           0.0145</span></span>
<span><span class="co">#&gt; 4 rec2                   0 CentralValleysp Deer_Cr_sp           0.0145</span></span>
<span><span class="co">#&gt; 5 rec2                   0 CentralValleysp UpperSacramento_R_sp 0.0145</span></span>
<span><span class="co">#&gt; 6 rec2                   0 CentralValleyfa Feather_H_fa         0.0145</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $bootstrapped_proportions</span></span>
<span><span class="co">#&gt; # A tibble: 0 × 0</span></span></code></pre></div>
<div class="section level2">
<h2 id="setting-the-prior-for-the-mixing-proportions">Setting the prior for the mixing proportions<a class="anchor" aria-label="anchor" href="#setting-the-prior-for-the-mixing-proportions"></a>
</h2>
<p>In some cases there might be a reason to explicitly set the parameters of the Dirichlet prior on the mixing proportions of the collections. For a contrived example, we could imagine that we wanted a Dirichlet prior with all parameters equal to 1/(# of collections), except for the parameters for all the Central Valley Fall Run populations, to which we would like to assign Dirichlet parameters of 2. That can be accomplished with the <code>pi_prior</code> argument to the <code><a href="reference/infer_mixture.html">infer_mixture()</a></code> function, which will let you pass in a tibble in which one column named “collection” gives the collection, and the other column, named “pi_param” gives the desired parameter.</p>
<p>Here we construct that kind of input:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prior_tibble</span> <span class="op">&lt;-</span> <span class="va">chinook</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">repunit</span>, <span class="va">collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op">==</span> <span class="st">"CentralValleyfa"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pi_param <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># see what it looks like:</span></span>
<span><span class="va">prior_tibble</span></span>
<span><span class="co">#&gt; # A tibble: 8 × 2</span></span>
<span><span class="co">#&gt;   collection      pi_param</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 Battle_Cr              2</span></span>
<span><span class="co">#&gt; 2 Butte_Cr_fa            2</span></span>
<span><span class="co">#&gt; 3 Deer_Cr_fa             2</span></span>
<span><span class="co">#&gt; 4 Feather_H_fa           2</span></span>
<span><span class="co">#&gt; 5 Feather_H_sp           2</span></span>
<span><span class="co">#&gt; 6 Mill_Cr_fa             2</span></span>
<span><span class="co">#&gt; 7 Mokelumne_R_fa         2</span></span>
<span><span class="co">#&gt; 8 Sacramento_R_lf        2</span></span></code></pre></div>
<p>Then we can run that in infer_mixture():</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12</span><span class="op">)</span></span>
<span><span class="va">mix_est_with_prior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                         mixture <span class="op">=</span> <span class="va">chinook_mix</span>, </span>
<span>                         gen_start_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                         pi_prior <span class="op">=</span> <span class="va">prior_tibble</span><span class="op">)</span></span>
<span><span class="co">#&gt; Collating data; compiling reference allele frequencies, etc.   time: 0.66 seconds</span></span>
<span><span class="co">#&gt; Computing reference locus specific means and variances for computing mixture z-scores   time: 0.10 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec2 with 772 individuals</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(collection)`  calculating log-likelihoods of the mixture individuals.   time: 0.07 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.55 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec1 with 743 individuals</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(collection)`  calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.52 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec3 with 741 individuals</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(collection)`  calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.52 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.03 seconds</span></span></code></pre></div>
<p>and now, for fun, we can compare the results for the mixing proportions of different collections there with and without the prior for the mixture collection rec1:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comp_mix_ests</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  `pi (default prior)` <span class="op">=</span> <span class="va">mix_est</span><span class="op">$</span><span class="va">mixing_proportions</span>,</span>
<span>  `pi (cv fall gets 2s prior)` <span class="op">=</span> <span class="va">mix_est_with_prior</span><span class="op">$</span><span class="va">mixing_proportions</span></span>
<span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="st">"prior_type"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">mixture_collection</span> <span class="op">==</span> <span class="st">"rec1"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">prior_type</span>, <span class="va">repunit</span>, <span class="va">collection</span>, <span class="va">pi</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/spread.html" class="external-link">spread</a></span><span class="op">(</span><span class="va">prior_type</span>, <span class="va">pi</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>coll_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op">==</span> <span class="st">"CentralValleyfa"</span>, <span class="st">"CV_fall"</span>, <span class="st">"Not_CV_fall"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comp_mix_ests</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`pi (default prior)`</span>, </span>
<span>           y <span class="op">=</span> <span class="va">`pi (cv fall gets 2s prior)`</span>,</span>
<span>           colour <span class="op">=</span> <span class="va">coll_group</span></span>
<span>           <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/unnamed-chunk-6-1.png"><!-- --></p>
<p>Yep, slightly different than before. Let’s look at the sums of everything:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comp_mix_ests</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">coll_group</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>with_explicit_prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">`pi (cv fall gets 2s prior)`</span><span class="op">)</span>, </span>
<span>            with_default_prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">`pi (default prior)`</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 3</span></span>
<span><span class="co">#&gt;   coll_group  with_explicit_prior with_default_prior</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                     &lt;dbl&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 CV_fall                   0.824              0.821</span></span>
<span><span class="co">#&gt; 2 Not_CV_fall               0.176              0.179</span></span></code></pre></div>
<p>We see that for the most part this change to the prior changed the distribution of fish into different collections within the Central Valley Fall reporting unit. This is not suprising—it is very hard to tell apart fish from those different collections. However, it did not greatly change the estimated proportion of the whole reporting unit. This also turns out to make sense if you consider the effect that the extra weight in the prior will have.</p>
</div>
<div class="section level2">
<h2 id="aggregating-collections-into-reporting-units">Aggregating collections into reporting units<a class="anchor" aria-label="anchor" href="#aggregating-collections-into-reporting-units"></a>
</h2>
<p>This is a simple operation in the <a href="https://tidyverse.org/" class="external-link">tidyverse</a>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># for mixing proportions</span></span>
<span><span class="va">rep_mix_ests</span> <span class="op">&lt;-</span> <span class="va">mix_est</span><span class="op">$</span><span class="va">mixing_proportions</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repprop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span>  <span class="co"># adding mixing proportions over collections in the repunit</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'mixture_collection'. You can override</span></span>
<span><span class="co">#&gt; using the `.groups` argument.</span></span>
<span></span>
<span><span class="co"># for individuals posteriors</span></span>
<span><span class="va">rep_indiv_ests</span> <span class="op">&lt;-</span> <span class="va">mix_est</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">indiv</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>rep_pofz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PofZ</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'mixture_collection', 'indiv'. You can</span></span>
<span><span class="co">#&gt; override using the `.groups` argument.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-posterior-density-curves-from-the-traces">Creating posterior density curves from the traces<a class="anchor" aria-label="anchor" href="#creating-posterior-density-curves-from-the-traces"></a>
</h2>
<p>The full MCMC output for the mixing proportions is available by default in the field <code>$mix_prop_traces</code>. This can be used to obtain an estimate of the posterior density of the mixing proportions.</p>
<p>Here we plot kernel density estimates for the 6 most abundant repunits from the <code>rec1</code> fishery:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find the top 6 most abundant:</span></span>
<span><span class="va">top6</span> <span class="op">&lt;-</span> <span class="va">rep_mix_ests</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">mixture_collection</span> <span class="op">==</span> <span class="st">"rec1"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">repprop</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check how many MCMC sweeps were done:</span></span>
<span><span class="va">nsweeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">mix_est</span><span class="op">$</span><span class="va">mix_prop_traces</span><span class="op">$</span><span class="va">sweep</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># keep only rec1, then discard the first 200 sweeps as burn-in,</span></span>
<span><span class="co"># and then aggregate over reporting units</span></span>
<span><span class="co"># and then keep only the top6 from above</span></span>
<span><span class="va">trace_subset</span> <span class="op">&lt;-</span> <span class="va">mix_est</span><span class="op">$</span><span class="va">mix_prop_traces</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">mixture_collection</span> <span class="op">==</span> <span class="st">"rec1"</span>, <span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">200</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repprop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">top6</span><span class="op">$</span><span class="va">repunit</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'sweep'. You can override using the</span></span>
<span><span class="co">#&gt; `.groups` argument.</span></span>
<span></span>
<span></span>
<span><span class="co"># now we can plot those:</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">trace_subset</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">repprop</span>, colour <span class="op">=</span> <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/plot-6-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="computing-credible-intervals-from-the-traces">Computing Credible Intervals from the Traces<a class="anchor" aria-label="anchor" href="#computing-credible-intervals-from-the-traces"></a>
</h2>
<p>Following on from the above example, we will use <code>trace_subset</code> to compute the equal-tail 95% credible intervals for the 6 most abundant reporting units in the <code>rec1</code> fishery:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top6_cis</span> <span class="op">&lt;-</span> <span class="va">trace_subset</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>loCI <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">repprop</span>, probs <span class="op">=</span> <span class="fl">0.025</span><span class="op">)</span>,</span>
<span>            hiCI <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">repprop</span>, probs <span class="op">=</span> <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top6_cis</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span><span class="co">#&gt;   repunit                     loCI   hiCI</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                      &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 CaliforniaCoast         1.86e- 2 0.0437</span></span>
<span><span class="co">#&gt; 2 CentralValleyfa         7.90e- 1 0.848 </span></span>
<span><span class="co">#&gt; 3 KlamathR                4.95e- 2 0.0865</span></span>
<span><span class="co">#&gt; 4 NCaliforniaSOregonCoast 3.12e- 3 0.0184</span></span>
<span><span class="co">#&gt; 5 RogueR                  4.36e- 2 0.0814</span></span>
<span><span class="co">#&gt; 6 UColumbiaRsufa          2.45e-18 0.0122</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="assessing-whether-individuals-are-not-from-any-of-the-reference-populations">Assessing whether individuals are not from any of the reference populations<a class="anchor" aria-label="anchor" href="#assessing-whether-individuals-are-not-from-any-of-the-reference-populations"></a>
</h2>
<p>Sometimes totally unexpected things happen. One situation we saw in the California Chinook fishery was samples coming to us that were actually coho salmon. Before we included coho salmon in the reference sample, these coho always assigned quite strongly to Alaska populations of Chinook, even though they don’t really look like Chinook at all.</p>
<p>In this case, it is useful to look at the raw log-likelihood values computed for the individual, rather than the scaled posterior probabilities. Because aberrantly low values of the genotype log-likelihood can indicate that there is something wrong. However, the raw likelihood that you get will depend on the number of missing loci, etc. <code>rubias</code> deals with this by computing a <em>z-score</em> for each fish. The Z-score is the Z statistic obtained from the fish’s log-likelihood (by subtracting from it the expected log-likelihood and dividing by the expected standard deviation). <code>rubias</code>’s implementation of the z-score accounts for the pattern of missing data, but it does this without all the simulation that <a href="https://github.com/eriqande/gsi_sim" class="external-link"><code>gsi_sim</code></a> does. This makes it much, much, faster—fast enough that we can compute it be default for every fish and every population.</p>
<p>Here, we will look at the z-score computed for each fish to the population with the highest posterior. (It is worth noting that you would <strong>never</strong> want to use the z-score to assign fish to different populations—it is only there to decide whether it looks like it might not have actually come from the population that it was assigned to, or any other population in the reference data set.)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the maximum-a-posteriori population for each individual</span></span>
<span><span class="va">map_rows</span> <span class="op">&lt;-</span> <span class="va">mix_est</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">indiv</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">PofZ</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If everything is kosher, then we expect that the z-scores we see will be roughly normally distributed. We can compare the distribution of z-scores we see with a bunch of simulated normal random variables.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>z_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1e06</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">map_rows</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">z_score</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">normo</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/z-score-density-1.png"><!-- --></p>
<p>The normal density is in black and the distribution of our observed z_scores is in blue. They fit reasonably well, suggesting that there is not too much weird stuff going on overall. (That is good!)</p>
<p>The z_score statistic is most useful as a check for individuals. It is intended to be a quick way to identify aberrant individuals. If you see a z-score to the maximum-a-posteriori population for an individual in your mixture sample that is considerably less than z_scores you saw in the reference, then you might infer that the individual doesn’t actually fit any of the populations in the reference well.</p>
</div>
<div class="section level2">
<h2 id="individuals-of-known-origin-in-the-mixture">Individuals of known origin in the mixture<a class="anchor" aria-label="anchor" href="#individuals-of-known-origin-in-the-mixture"></a>
</h2>
<p>Here I include a small, contrived example. We use the <code>small_chinook</code> data set so that it goes fast.</p>
<p>First, we analyze the data with no fish in the mixture of known collection</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">no_kc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span><span class="va">small_chinook_ref</span>, <span class="va">small_chinook_mix</span>, gen_start_col <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Collating data; compiling reference allele frequencies, etc.   time: 0.09 seconds</span></span>
<span><span class="co">#&gt; Computing reference locus specific means and variances for computing mixture z-scores   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec3 with 29 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec1 with 36 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec2 with 35 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span></code></pre></div>
<p>And look at the results for the mixing proportions:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">no_kc</span><span class="op">$</span><span class="va">mixing_proportions</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 18 × 4</span></span>
<span><span class="co">#&gt;    mixture_collection repunit         collection          pi</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;           &lt;chr&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 rec1               CentralValleyfa Feather_H_fa   0.849  </span></span>
<span><span class="co">#&gt;  2 rec1               CentralValleysp Deer_Cr_sp     0.0508 </span></span>
<span><span class="co">#&gt;  3 rec1               CaliforniaCoast Eel_R          0.0400 </span></span>
<span><span class="co">#&gt;  4 rec1               KlamathR        Klamath_IGH_fa 0.0305 </span></span>
<span><span class="co">#&gt;  5 rec1               MidOregonCoast  Umpqua_sp      0.0252 </span></span>
<span><span class="co">#&gt;  6 rec1               CentralValleywi Sacramento_H   0.00463</span></span>
<span><span class="co">#&gt;  7 rec2               CentralValleyfa Feather_H_fa   0.809  </span></span>
<span><span class="co">#&gt;  8 rec2               KlamathR        Klamath_IGH_fa 0.103  </span></span>
<span><span class="co">#&gt;  9 rec2               MidOregonCoast  Umpqua_sp      0.0733 </span></span>
<span><span class="co">#&gt; 10 rec2               CentralValleysp Deer_Cr_sp     0.00550</span></span>
<span><span class="co">#&gt; 11 rec2               CaliforniaCoast Eel_R          0.00479</span></span>
<span><span class="co">#&gt; 12 rec2               CentralValleywi Sacramento_H   0.00474</span></span>
<span><span class="co">#&gt; 13 rec3               CentralValleyfa Feather_H_fa   0.839  </span></span>
<span><span class="co">#&gt; 14 rec3               CaliforniaCoast Eel_R          0.0714 </span></span>
<span><span class="co">#&gt; 15 rec3               MidOregonCoast  Umpqua_sp      0.0496 </span></span>
<span><span class="co">#&gt; 16 rec3               KlamathR        Klamath_IGH_fa 0.0262 </span></span>
<span><span class="co">#&gt; 17 rec3               CentralValleysp Deer_Cr_sp     0.00875</span></span>
<span><span class="co">#&gt; 18 rec3               CentralValleywi Sacramento_H   0.00542</span></span></code></pre></div>
<p>Now, we will do the same analysis, but pretend that we know that the first 8 of the 36 fish in fishery rec1 are from the Deer_Cr_sp collection.</p>
<p>First we have to add the known_collection column to the reference.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make reference file that includes the known_collection column</span></span>
<span><span class="va">kc_ref</span> <span class="op">&lt;-</span> <span class="va">small_chinook_ref</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>known_collection <span class="op">=</span> <span class="va">collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">known_collection</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># see what that looks like</span></span>
<span><span class="va">kc_ref</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">#&gt; # A tibble: 10 × 8</span></span>
<span><span class="co">#&gt;    known_collection sample_type repunit         collection indiv   Ots_94857.232</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt;  2 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt;  3 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt;  4 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             4</span></span>
<span><span class="co">#&gt;  5 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt;  6 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             4</span></span>
<span><span class="co">#&gt;  7 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt;  8 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             4</span></span>
<span><span class="co">#&gt;  9 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt; 10 Deer_Cr_sp       reference   CentralValleysp Deer_Cr_sp Deer_C…             2</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: Ots_94857.232.1 &lt;int&gt;, Ots_102213.210 &lt;int&gt;</span></span></code></pre></div>
<p>Then we add the known collection column to the mixture. We start out making it all NAs, and then we change that to Deer_Cr_sp for 8 of the rec1 fish:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kc_mix</span> <span class="op">&lt;-</span> <span class="va">small_chinook_mix</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>known_collection <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">known_collection</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kc_mix</span><span class="op">$</span><span class="va">known_collection</span><span class="op">[</span><span class="va">kc_mix</span><span class="op">$</span><span class="va">collection</span> <span class="op">==</span> <span class="st">"rec1"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Deer_Cr_sp"</span></span>
<span></span>
<span><span class="co"># here is what that looks like now (dropping most of the genetic data columns)</span></span>
<span><span class="va">kc_mix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">#&gt; # A tibble: 20 × 7</span></span>
<span><span class="co">#&gt;    known_collection sample_type repunit collection indiv   Ots_94857.232</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T125347             4</span></span>
<span><span class="co">#&gt;  2 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T127759             4</span></span>
<span><span class="co">#&gt;  3 &lt;NA&gt;             mixture     &lt;NA&gt;    rec2       T124955             4</span></span>
<span><span class="co">#&gt;  4 &lt;NA&gt;             mixture     &lt;NA&gt;    rec2       T127564             2</span></span>
<span><span class="co">#&gt;  5 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T127392             4</span></span>
<span><span class="co">#&gt;  6 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T127414             4</span></span>
<span><span class="co">#&gt;  7 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T124839             4</span></span>
<span><span class="co">#&gt;  8 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T126414             2</span></span>
<span><span class="co">#&gt;  9 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T125252             4</span></span>
<span><span class="co">#&gt; 10 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T127765             4</span></span>
<span><span class="co">#&gt; 11 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T127293             2</span></span>
<span><span class="co">#&gt; 12 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T127577             2</span></span>
<span><span class="co">#&gt; 13 &lt;NA&gt;             mixture     &lt;NA&gt;    rec2       T126766             4</span></span>
<span><span class="co">#&gt; 14 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T126494             4</span></span>
<span><span class="co">#&gt; 15 &lt;NA&gt;             mixture     &lt;NA&gt;    rec2       T125205             4</span></span>
<span><span class="co">#&gt; 16 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T126584             4</span></span>
<span><span class="co">#&gt; 17 &lt;NA&gt;             mixture     &lt;NA&gt;    rec2       T124821             4</span></span>
<span><span class="co">#&gt; 18 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T124905             2</span></span>
<span><span class="co">#&gt; 19 Deer_Cr_sp       mixture     &lt;NA&gt;    rec1       T124735             4</span></span>
<span><span class="co">#&gt; 20 &lt;NA&gt;             mixture     &lt;NA&gt;    rec3       T125624             2</span></span>
<span><span class="co">#&gt; # ℹ 1 more variable: Ots_94857.232.1 &lt;int&gt;</span></span></code></pre></div>
<p>And now we can do the mixture analysis:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note that the genetic data start in column 6 now</span></span>
<span><span class="va">with_kc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span><span class="va">kc_ref</span>, <span class="va">kc_mix</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Collating data; compiling reference allele frequencies, etc.   time: 0.09 seconds</span></span>
<span><span class="co">#&gt; Computing reference locus specific means and variances for computing mixture z-scores   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec3 with 29 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec1 with 36 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span>
<span><span class="co">#&gt; Working on mixture collection: rec2 with 35 individuals</span></span>
<span><span class="co">#&gt;   calculating log-likelihoods of the mixture individuals.   time: 0.00 seconds</span></span>
<span><span class="co">#&gt;   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.02 seconds</span></span>
<span><span class="co">#&gt;   tidying output into a tibble.   time: 0.01 seconds</span></span></code></pre></div>
<p>And, when we look at the estimated proportions, we see that for rec1 they reflect the fact that 8 of those fish were singled out as known fish from Deer_Ck_sp:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">with_kc</span><span class="op">$</span><span class="va">mixing_proportions</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 18 × 4</span></span>
<span><span class="co">#&gt;    mixture_collection repunit         collection          pi</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;           &lt;chr&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 rec1               CentralValleyfa Feather_H_fa   0.546  </span></span>
<span><span class="co">#&gt;  2 rec1               CentralValleysp Deer_Cr_sp     0.355  </span></span>
<span><span class="co">#&gt;  3 rec1               CaliforniaCoast Eel_R          0.0411 </span></span>
<span><span class="co">#&gt;  4 rec1               KlamathR        Klamath_IGH_fa 0.0318 </span></span>
<span><span class="co">#&gt;  5 rec1               MidOregonCoast  Umpqua_sp      0.0220 </span></span>
<span><span class="co">#&gt;  6 rec1               CentralValleywi Sacramento_H   0.00438</span></span>
<span><span class="co">#&gt;  7 rec2               CentralValleyfa Feather_H_fa   0.806  </span></span>
<span><span class="co">#&gt;  8 rec2               KlamathR        Klamath_IGH_fa 0.104  </span></span>
<span><span class="co">#&gt;  9 rec2               MidOregonCoast  Umpqua_sp      0.0732 </span></span>
<span><span class="co">#&gt; 10 rec2               CentralValleysp Deer_Cr_sp     0.00688</span></span>
<span><span class="co">#&gt; 11 rec2               CaliforniaCoast Eel_R          0.00551</span></span>
<span><span class="co">#&gt; 12 rec2               CentralValleywi Sacramento_H   0.00456</span></span>
<span><span class="co">#&gt; 13 rec3               CentralValleyfa Feather_H_fa   0.835  </span></span>
<span><span class="co">#&gt; 14 rec3               CaliforniaCoast Eel_R          0.0732 </span></span>
<span><span class="co">#&gt; 15 rec3               MidOregonCoast  Umpqua_sp      0.0494 </span></span>
<span><span class="co">#&gt; 16 rec3               KlamathR        Klamath_IGH_fa 0.0281 </span></span>
<span><span class="co">#&gt; 17 rec3               CentralValleysp Deer_Cr_sp     0.00883</span></span>
<span><span class="co">#&gt; 18 rec3               CentralValleywi Sacramento_H   0.00565</span></span></code></pre></div>
<p>The output from <code><a href="reference/infer_mixture.html">infer_mixture()</a></code> in this case can be used just like it was before without known individuals in the baseline.</p>
</div>
<div class="section level2">
<h2 id="fully-bayesian-model-with-updating-of-allele-freqencies">Fully Bayesian model (with updating of allele freqencies)<a class="anchor" aria-label="anchor" href="#fully-bayesian-model-with-updating-of-allele-freqencies"></a>
</h2>
<p>The default model in <code>rubias</code> is a conditional model in which inference is done with the baseline allele counts fixed. In a fully Bayesian version, fish from within the mixture that are allocated (on any particular step of the MCMC) to one of the reference samples have their alleles added to that reference sample, thus (one hopes) refining the estimate of allele frequencies in that sample. This is more computationally intensive, and, is done using parallel computation, by default running one thread for every core on your machine.</p>
<p>The basic way to invoke the fully Bayesian model is to use the <code>infer_mixture</code> function with the <code>method</code> option set to “BR”. For example:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_model_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>  mixture <span class="op">=</span> <span class="va">chinook_mix</span>, </span>
<span>  gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  method <span class="op">=</span> <span class="st">"BR"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>More details about different options for working with the fully Bayesian model are available in the vignette about the fully Bayesian model.</p>
</div>
</div>
<div class="section level1">
<h1 id="assessment-of-genetic-references">Assessment of Genetic References<a class="anchor" aria-label="anchor" href="#assessment-of-genetic-references"></a>
</h1>
<div class="section level2">
<h2 id="self-assigning-fish-from-the-reference">Self-assigning fish from the reference<a class="anchor" aria-label="anchor" href="#self-assigning-fish-from-the-reference"></a>
</h2>
<p>A standard analysis in molecular ecology is to assign individuals in the reference back to the collections in the reference using a <em>leave-one-out</em> procedure. This is taken care of by the <code><a href="reference/self_assign.html">self_assign()</a></code> function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sa_chinook</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/self_assign.html">self_assign</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, gen_start_col <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Summary Statistics:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 7301 Individuals in Sample</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 91 Loci: AldB1.122, AldoB4.183, OTNAML12_1.SNP1, Ots_100884.287, Ots_101119.381, Ots_101704.143, Ots_102213.210, Ots_102414.395, Ots_102420.494, Ots_102457.132, Ots_102801.308, Ots_102867.609, Ots_103041.52, Ots_104063.132, Ots_104569.86, Ots_105105.613, Ots_105132.200, Ots_105401.325, Ots_105407.117, Ots_106499.70, Ots_106747.239, Ots_107074.284, Ots_107285.93, Ots_107806.821, Ots_108007.208, Ots_108390.329, Ots_108735.302, Ots_109693.392, Ots_110064.383, Ots_110201.363, Ots_110495.380, Ots_110551.64, Ots_111312.435, Ots_111666.408, Ots_111681.657, Ots_112301.43, Ots_112419.131, Ots_112820.284, Ots_112876.371, Ots_113242.216, Ots_113457.40, Ots_117043.255, Ots_117242.136, Ots_117432.409, Ots_118175.479, Ots_118205.61, Ots_118938.325, Ots_122414.56, Ots_123048.521, Ots_123921.111, Ots_124774.477, Ots_127236.62, Ots_128302.57, Ots_128693.461, Ots_128757.61, Ots_129144.472, Ots_129170.683, Ots_129458.451, Ots_130720.99, Ots_131460.584, Ots_131906.141, Ots_94857.232, Ots_96222.525, Ots_96500.180, Ots_97077.179, Ots_99550.204, Ots_ARNT.195, Ots_AsnRS.60, Ots_aspat.196, Ots_CD59.2, Ots_CD63, Ots_EP.529, Ots_GDH.81x, Ots_HSP90B.385, Ots_MHC1, Ots_mybp.85, Ots_myoD.364, Ots_Ots311.101x, Ots_PGK.54, Ots_Prl2, Ots_RFC2.558, Ots_SClkF2R2.135, Ots_SWS1op.182, Ots_TAPBP, Ots_u07.07.161, Ots_u07.49.290, Ots_u4.92, OTSBMP.2.SNP1, OTSTF1.SNP1, S71.336, unk_526</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 39 Reporting Units: CentralValleyfa, CentralValleysp, CentralValleywi, CaliforniaCoast, KlamathR, NCaliforniaSOregonCoast, RogueR, MidOregonCoast, NOregonCoast, WillametteR, DeschutesRfa, LColumbiaRfa, LColumbiaRsp, MidColumbiaRtule, UColumbiaRsufa, MidandUpperColumbiaRsp, SnakeRfa, SnakeRspsu, NPugetSound, WashingtonCoast, SPugetSound, LFraserR, LThompsonR, EVancouverIs, WVancouverIs, MSkeenaR, MidSkeenaR, LSkeenaR, SSEAlaska, NGulfCoastAlsekR, NGulfCoastKarlukR, TakuR, NSEAlaskaChilkatR, NGulfCoastSitukR, CopperR, SusitnaR, LKuskokwimBristolBay, MidYukon, CohoSp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 69 Collections: Feather_H_sp, Butte_Cr_Sp, Mill_Cr_sp, Deer_Cr_sp, UpperSacramento_R_sp, Feather_H_fa, Butte_Cr_fa, Mill_Cr_fa, Deer_Cr_fa, Mokelumne_R_fa, Battle_Cr, Sacramento_R_lf, Sacramento_H, Eel_R, Russian_R, Klamath_IGH_fa, Trinity_H_sp, Smith_R, Chetco_R, Cole_Rivers_H, Applegate_Cr, Coquille_R, Umpqua_sp, Nestucca_H, Siuslaw_R, Alsea_R, Nehalem_R, Siletz_R, N_Santiam_H, McKenzie_H, L_Deschutes_R, Cowlitz_H_fa, Cowlitz_H_sp, Kalama_H_sp, Spring_Cr_H, Hanford_Reach, PriestRapids_H, Wells_H, Wenatchee_R, CleElum, Lyons_Ferry_H, Rapid_R_H, McCall_H, Kendall_H_sp, Forks_Cr_H, Soos_H, Marblemount_H_sp, QuinaltLake_f, Harris_R, Birkenhead_H, Spius_H, Big_Qual_H, Robertson_H, Morice_R, Kitwanga_R, L_Kalum_R, LPW_Unuk_R, Goat_Cr, Karluk_R, LittleTatsamenie, Tahini_R, Situk_R, Sinona_Ck, Montana_Ck, George_R, Kanektok_R, Togiak_R, Kantishna_R, California_Coho</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4.18% of allelic data identified as missing</span></span></code></pre></div>
<p>Now, you can look at the self assignment results:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sa_chinook</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 100 × 11</span></span>
<span><span class="co">#&gt;    indiv             collection   repunit   inferred_collection inferred_repunit</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;             &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;               &lt;chr&gt;           </span></span>
<span><span class="co">#&gt;  1 Feather_H_sp:0001 Feather_H_sp CentralV… Feather_H_sp        CentralValleyfa </span></span>
<span><span class="co">#&gt;  2 Feather_H_sp:0001 Feather_H_sp CentralV… Feather_H_fa        CentralValleyfa </span></span>
<span><span class="co">#&gt;  3 Feather_H_sp:0001 Feather_H_sp CentralV… Butte_Cr_fa         CentralValleyfa </span></span>
<span><span class="co">#&gt;  4 Feather_H_sp:0001 Feather_H_sp CentralV… Mill_Cr_sp          CentralValleysp </span></span>
<span><span class="co">#&gt;  5 Feather_H_sp:0001 Feather_H_sp CentralV… Mill_Cr_fa          CentralValleyfa </span></span>
<span><span class="co">#&gt;  6 Feather_H_sp:0001 Feather_H_sp CentralV… UpperSacramento_R_… CentralValleysp </span></span>
<span><span class="co">#&gt;  7 Feather_H_sp:0001 Feather_H_sp CentralV… Deer_Cr_sp          CentralValleysp </span></span>
<span><span class="co">#&gt;  8 Feather_H_sp:0001 Feather_H_sp CentralV… Butte_Cr_Sp         CentralValleysp </span></span>
<span><span class="co">#&gt;  9 Feather_H_sp:0001 Feather_H_sp CentralV… Battle_Cr           CentralValleyfa </span></span>
<span><span class="co">#&gt; 10 Feather_H_sp:0001 Feather_H_sp CentralV… Mokelumne_R_fa      CentralValleyfa </span></span>
<span><span class="co">#&gt; # ℹ 90 more rows</span></span>
<span><span class="co">#&gt; # ℹ 6 more variables: scaled_likelihood &lt;dbl&gt;, log_likelihood &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   z_score &lt;dbl&gt;, n_non_miss_loci &lt;int&gt;, n_miss_loci &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   missing_loci &lt;list&gt;</span></span></code></pre></div>
<p>The <code>log_likelihood</code> is the log probability of the fish’s genotype given it is from the <code>inferred_collection</code> computed using leave-one-out. The <code>scaled_likelihood</code> is the posterior prob of assigning the fish to the <code>inferred_collection</code> given an equal prior on every collection in the reference. Other columns are as in the output for <code><a href="reference/infer_mixture.html">infer_mixture()</a></code>. Note that the <code>z_score</code> computed here can be used to assess the distribution of the <code>z_score</code> statistic for fish from known, reference populations. This can be used to compare to values obtained in mixed fisheries.</p>
<p>The output can be summarized by repunit as was done above:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sa_to_repu</span> <span class="op">&lt;-</span> <span class="va">sa_chinook</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">indiv</span>, <span class="va">collection</span>, <span class="va">repunit</span>, <span class="va">inferred_repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repu_scaled_like <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">scaled_likelihood</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'indiv', 'collection', 'repunit'. You can</span></span>
<span><span class="co">#&gt; override using the `.groups` argument.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sa_to_repu</span>, n <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 200 × 5</span></span>
<span><span class="co">#&gt; # Groups:   indiv, collection, repunit [6]</span></span>
<span><span class="co">#&gt;    indiv        collection repunit      inferred_repunit repu_scaled_like</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;                       &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 Alsea_R:0001 Alsea_R    NOregonCoast CaliforniaCoast          3.72e- 8</span></span>
<span><span class="co">#&gt;  2 Alsea_R:0001 Alsea_R    NOregonCoast CentralValleyfa          1.54e-14</span></span>
<span><span class="co">#&gt;  3 Alsea_R:0001 Alsea_R    NOregonCoast CentralValleysp          8.12e-15</span></span>
<span><span class="co">#&gt;  4 Alsea_R:0001 Alsea_R    NOregonCoast CentralValleywi          1.22e-23</span></span>
<span><span class="co">#&gt;  5 Alsea_R:0001 Alsea_R    NOregonCoast CohoSp                   2.09e-52</span></span>
<span><span class="co">#&gt;  6 Alsea_R:0001 Alsea_R    NOregonCoast CopperR                  3.08e-20</span></span>
<span><span class="co">#&gt;  7 Alsea_R:0001 Alsea_R    NOregonCoast DeschutesRfa             3.81e-10</span></span>
<span><span class="co">#&gt;  8 Alsea_R:0001 Alsea_R    NOregonCoast EVancouverIs             1.02e- 8</span></span>
<span><span class="co">#&gt;  9 Alsea_R:0001 Alsea_R    NOregonCoast KlamathR                 1.11e-11</span></span>
<span><span class="co">#&gt; 10 Alsea_R:0001 Alsea_R    NOregonCoast LColumbiaRfa             8.52e- 8</span></span>
<span><span class="co">#&gt; # ℹ 190 more rows</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulated-mixtures-using-a-leave-one-out-type-of-approach">Simulated mixtures using a leave-one-out type of approach<a class="anchor" aria-label="anchor" href="#simulated-mixtures-using-a-leave-one-out-type-of-approach"></a>
</h2>
<p>If you want to know how much accuracy you can expect given a set of genetic markers and a grouping of populations (<code>collection</code>s) into reporting units (<code>repunit</code>s), there are two different functions you might use:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="reference/assess_reference_loo.html">assess_reference_loo()</a></code>: This function carries out simulation of mixtures using the leave-one-out approach of Anderson, Waples, and Kalinowski (2008).</li>
<li>
<code><a href="reference/assess_reference_mc.html">assess_reference_mc()</a></code>: This functions breaks the reference data set into different subsets, one of which is used as the reference data set and the other the mixture. It is difficult to simulate very large mixture samples using this method, because it is constrained by the number of fish in the reference data set.<br>
Additionally, there are constraints on the mixing proportions that can be simulated because of variation in the number of fish from each collection in the reference.</li>
</ol>
<p>Both of the functions take two required arguments: 1) a data frame of reference genetic data, and 2) the number of the column in which the genetic data start.</p>
<p>Here we use the <code>chinook</code> data to simulate 50 mixture samples of size 200 fish using the default values (Dirichlet parameters of 1.5 for each reporting unit, and Dirichlet parameters of 1.5 for each collection within a reporting unit…)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chin_sims</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p>Here is what the output looks like:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chin_sims</span></span>
<span><span class="co">#&gt; # A tibble: 3,450 × 9</span></span>
<span><span class="co">#&gt;    repunit_scenario collection_scenario  iter repunit   collection true_pi     n</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;               &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 1                1                       1 CentralV… Feather_H… 8.42e-4     0</span></span>
<span><span class="co">#&gt;  2 1                1                       1 CentralV… Butte_Cr_… 6.55e-4     0</span></span>
<span><span class="co">#&gt;  3 1                1                       1 CentralV… Mill_Cr_sp 1.37e-3     0</span></span>
<span><span class="co">#&gt;  4 1                1                       1 CentralV… Deer_Cr_sp 4.41e-3     1</span></span>
<span><span class="co">#&gt;  5 1                1                       1 CentralV… UpperSacr… 6.25e-4     0</span></span>
<span><span class="co">#&gt;  6 1                1                       1 CentralV… Feather_H… 2.89e-3     2</span></span>
<span><span class="co">#&gt;  7 1                1                       1 CentralV… Butte_Cr_… 8.50e-4     0</span></span>
<span><span class="co">#&gt;  8 1                1                       1 CentralV… Mill_Cr_fa 2.86e-3     1</span></span>
<span><span class="co">#&gt;  9 1                1                       1 CentralV… Deer_Cr_fa 6.53e-3     0</span></span>
<span><span class="co">#&gt; 10 1                1                       1 CentralV… Mokelumne… 8.99e-4     0</span></span>
<span><span class="co">#&gt; # ℹ 3,440 more rows</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: post_mean_pi &lt;dbl&gt;, mle_pi &lt;dbl&gt;</span></span></code></pre></div>
<p>The columns here are:</p>
<ul>
<li>
<code>repunit_scenario</code> and integer that gives that repunit simulation parameters (see below about simulating multiple scenarios).</li>
<li>
<code>collections_scenario</code> and integer that gives that collection simulation paramters (see below about simulating multiple scenarios).</li>
<li>
<code>iter</code> the simulation number (1 up to <code>reps</code>)</li>
<li>
<code>repunit</code> the reporting unit</li>
<li>
<code>collection</code> the collection</li>
<li>
<code>true_pi</code> the true simulated mixing proportion</li>
<li>
<code>n</code> the actual number of fish from the collection in the simulated mixture.</li>
<li>
<code>post_mean_pi</code> the posterior mean of mixing proportion.</li>
<li>
<code>mle_pi</code> the maximum likelihood of <code>pi</code> obtained using an EM-algorithm.</li>
</ul>
</div>
<div class="section level2">
<h2 id="specifying-mixture-proportions-in-assess_reference_loo">Specifying mixture proportions in <code>assess_reference_loo()</code>
<a class="anchor" aria-label="anchor" href="#specifying-mixture-proportions-in-assess_reference_loo"></a>
</h2>
<p>By default, each iteration, the proportions of fish from each reporting unit are simulated from a Dirichlet distribution with parameter (1.5,…,1.5). And, within each reporting unit the mixing proportions from different collections are drawn from a Dirichlet distribution with parameter (1.5,…,1.5).</p>
<p>The value of 1.5 for the Dirichlet parameter for reporting units can be changed using the <code>alpha_repunit</code>. The Dirichlet parameter for collections can be set using the <code>alpha_collection</code> parameter.</p>
<p>Sometimes, however, more control over the composition of the simulated mixtures is desired. This is achieved by passing a two-column <em>data.frame</em> to either <code>alpha_repunit</code> or <code>alpha_collection</code> (or both). If you are passing the data.frame in for <code>alpha_repunit</code>, the first column must be named <code>repunit</code> and it must contain a character vector specifying reporting units. In the data.frame for <code>alpha_collection</code> the first column must be named <code>collection</code> and must hold a character vector specifying different collections. It is an error if a repunit or collection is specified that does not exist in the reference. However, you do not need to specify a value for every reporting unit or collection. (If they are absent, the value is assumed to be zero.)</p>
<p>The second column of the data frame must be one of <code>count</code>, <code>ppn</code> or <code>dirichlet</code>. These specify, respectively,</p>
<ol style="list-style-type: decimal">
<li>the exact count of individuals to be simulated from each repunit (or collection);</li>
<li>the proportion of individuals from each repunit (or collection). These <code>ppn</code> values will be normalized to sum to one if they do not. As such, they can be regarded as weights.</li>
<li>the parameters of a Dirichlet distribution from which the proportion of individuals should be simulated.</li>
</ol>
<p>Let’s say that we want to simulate data that roughly have proportions like what we saw in the Chinook <code>rec1</code> fishery. We have those estimates in the variable <code>top6</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top6</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span><span class="co">#&gt; # Groups:   mixture_collection [1]</span></span>
<span><span class="co">#&gt;   mixture_collection repunit                 repprop</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;                     &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 rec1               CentralValleyfa         0.821  </span></span>
<span><span class="co">#&gt; 2 rec1               KlamathR                0.0669 </span></span>
<span><span class="co">#&gt; 3 rec1               RogueR                  0.0606 </span></span>
<span><span class="co">#&gt; 4 rec1               CaliforniaCoast         0.0300 </span></span>
<span><span class="co">#&gt; 5 rec1               NCaliforniaSOregonCoast 0.00923</span></span>
<span><span class="co">#&gt; 6 rec1               UColumbiaRsufa          0.00430</span></span></code></pre></div>
<p>We could, if we put those <code>repprop</code> values into a <code>ppn</code> column, simulate mixtures with exactly those proportions. Or if we wanted to simulate exact numbers of fish in a sample of 347 fish, we could get those values like this:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">top6</span><span class="op">$</span><span class="va">repprop</span> <span class="op">*</span> <span class="fl">350</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 287  23  21  11   3   2</span></span></code></pre></div>
<p>and then put them in a <code>cnts</code> column.</p>
<p>However, in this case, we want to simulate mixtures that look similar to the one we estimated, but have some variation. For that we will want to supply Dirichlet random variable paramaters in a column named <code>dirichlet</code>. If we make the values proportional to the mixing proportions, then, on average that is what they will be. If the values are large, then there will be little variation between simulated mixtures. And if the the values are small there will be lots of variation. We’ll scale them so that they sum to 10—that should give some variation, but not too much. Accordingly the tibble that we pass in as the <code>alpha_repunit</code> parameter, which describes the variation in reporting unit proportions we would like to simulate would look like this:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arep</span> <span class="op">&lt;-</span> <span class="va">top6</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dirichlet <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="va">repprop</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">repunit</span>, <span class="va">dirichlet</span><span class="op">)</span></span>
<span></span>
<span><span class="va">arep</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   repunit                 dirichlet</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                       &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 CentralValleyfa            8.21  </span></span>
<span><span class="co">#&gt; 2 KlamathR                   0.669 </span></span>
<span><span class="co">#&gt; 3 RogueR                     0.606 </span></span>
<span><span class="co">#&gt; 4 CaliforniaCoast            0.300 </span></span>
<span><span class="co">#&gt; 5 NCaliforniaSOregonCoast    0.0923</span></span>
<span><span class="co">#&gt; 6 UColumbiaRsufa             0.0430</span></span></code></pre></div>
<p>Let’s do some simulations with those repunit parameters. By default, if we don’t specify anything extra for the <em>collections</em>, they get dirichlet parameters of 1.5.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chin_sims_repu_top6</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                     alpha_repunit <span class="op">=</span> <span class="va">arep</span><span class="op">)</span></span></code></pre></div>
<p>Now, we can summarise the output by reporting unit…</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># now, call those repunits that we did not specify in arep "OTHER"</span></span>
<span><span class="co"># and then sum up over reporting units</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">chin_sims_repu_top6</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>repunit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">arep</span><span class="op">$</span><span class="va">repunit</span>, <span class="va">repunit</span>, <span class="st">"OTHER"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>true_repprop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">true_pi</span><span class="op">)</span>, </span>
<span>            reprop_posterior_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">post_mean_pi</span><span class="op">)</span>,</span>
<span>            repu_n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>repu_n_prop <span class="op">=</span> <span class="va">repu_n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">repu_n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'iter'. You can override using the</span></span>
<span><span class="co">#&gt; `.groups` argument.</span></span></code></pre></div>
<p>…and then plot it for the values we are interested in:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># then plot them</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">true_repprop</span>, y <span class="op">=</span> <span class="va">reprop_posterior_mean</span>, colour <span class="op">=</span> <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span>, slope <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">repunit</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/plot-top6-1.png"><!-- --></p>
<p>Or plot comparing to their “n” value, which is the actual number of fish from each reporting unit in the sample.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">repu_n_prop</span>, y <span class="op">=</span> <span class="va">reprop_posterior_mean</span>, colour <span class="op">=</span> <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span>, slope <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">repunit</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/plot-top6-n-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="retrieving-the-individual-simulated-fish-posteriors">Retrieving the individual simulated fish posteriors<a class="anchor" aria-label="anchor" href="#retrieving-the-individual-simulated-fish-posteriors"></a>
</h2>
<p>Quite often you might be curious about how much you can expect to be able to trust the posterior for individual fish from a mixture like this. You can retrieve all the posteriors computed for the fish simulated in <code><a href="reference/assess_reference_loo.html">assess_reference_loo()</a></code> using the <code>return_indiv_posteriors</code> option. When you do this, the function returns a list with components <code>mixture_proportions</code> (which holds a tibble like <code>chin_sims_repu_top6</code> in the previous section) and <code>indiv_posteriors</code>, which holds all the posteriors (<code>PofZ</code>s) for the simulated individuals.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">chin_sims_with_indivs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                     alpha_repunit <span class="op">=</span> <span class="va">arep</span>,</span>
<span>                     return_indiv_posteriors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `as.tibble()` was deprecated in tibble 2.0.0.</span></span>
<span><span class="co">#&gt; ℹ Please use `as_tibble()` instead.</span></span>
<span><span class="co">#&gt; ℹ The signature and semantics have changed, see `?as_tibble`.</span></span>
<span><span class="co">#&gt; ℹ The deprecated feature was likely used in the rubias package.</span></span>
<span><span class="co">#&gt;   Please report the issue at &lt;https://github.com/eriqande/rubias/issues&gt;.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span>
<span></span>
<span><span class="co"># print out the indiv posteriors</span></span>
<span><span class="va">chin_sims_with_indivs</span><span class="op">$</span><span class="va">indiv_posteriors</span></span>
<span><span class="co">#&gt; # A tibble: 690,000 × 9</span></span>
<span><span class="co">#&gt;    repunit_scenario collection_scenario  iter indiv simulated_repunit</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;               &lt;int&gt; &lt;int&gt; &lt;chr&gt;            </span></span>
<span><span class="co">#&gt;  1 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  2 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  3 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  4 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  5 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  6 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  7 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  8 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt;  9 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt; 10 1                1                       1     1 CentralValleyfa  </span></span>
<span><span class="co">#&gt; # ℹ 689,990 more rows</span></span>
<span><span class="co">#&gt; # ℹ 4 more variables: simulated_collection &lt;chr&gt;, repunit &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   collection &lt;chr&gt;, PofZ &lt;dbl&gt;</span></span></code></pre></div>
<p>In this tibble: - <code>indiv</code> is an integer specifier of the simulated individual - <code>simulated_repunit</code> is the reporting unit the individual was simulated from - <code>simulated_collection</code> is the collection the simulated genotype came from - <code>PofZ</code> is the mean over the MCMC of the posterior probability that the individual originated from the collection.</p>
<p>Now that we have done that, we can see what the distribution of posteriors to the correct reporting unit is for fish from the different simulated collections. We’ll do that with a boxplot, coloring by repunit:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarise things</span></span>
<span><span class="va">repu_pofzs</span> <span class="op">&lt;-</span> <span class="va">chin_sims_with_indivs</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op">==</span> <span class="va">simulated_repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">indiv</span>, <span class="va">simulated_collection</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>  <span class="co"># first aggregate over reporting units</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repu_PofZ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PofZ</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">repunit</span>, <span class="va">simulated_collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulated_collection <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">simulated_collection</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simulated_collection</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'iter', 'indiv', 'simulated_collection'.</span></span>
<span><span class="co">#&gt; You can override using the `.groups` argument.</span></span>
<span></span>
<span><span class="co"># also get the number of simulated individuals from each collection</span></span>
<span><span class="va">num_simmed</span> <span class="op">&lt;-</span> <span class="va">chin_sims_with_indivs</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">indiv</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">simulated_collection</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># note, the last few steps make simulated collection a factor so that collections within</span></span>
<span><span class="co"># the same repunit are grouped together in the plot.</span></span>
<span></span>
<span><span class="co"># now, plot it</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">repu_pofzs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">simulated_collection</span>, y <span class="op">=</span> <span class="va">repu_PofZ</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">num_simmed</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fl">1.025</span>, label <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">0</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">9</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">1.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/boxplot-pofz-indiv-sim-1.png"><!-- --></p>
<p>Great. That is helpful.</p>
</div>
<div class="section level2">
<h2 id="changing-the-resampling-unit">Changing the resampling unit<a class="anchor" aria-label="anchor" href="#changing-the-resampling-unit"></a>
</h2>
<p>By default, individuals are simulated in <code><a href="reference/assess_reference_loo.html">assess_reference_loo()</a></code> by resampling full multilocus genotypes. This tends to be more realistic, because it includes as missing in the simulations all the missing data for individuals in the reference. However, as all the genes in individuals that have been incorrectly placed in a reference stay together, that individual might have a low value of PofZ to the population it was simulated from. Due to the latter issue, it might also yield a more pessimistic assessment’ of the power for GSI.</p>
<p>An alternative is to resample over gene copies—the CV-GC method of Anderson, Waples, and Kalinowski (2008).</p>
<p>Let us do that and see how the simulated PofZ results change. Here we do the simulations…</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">101</span><span class="op">)</span> <span class="co"># for reproducibility</span></span>
<span><span class="co"># do the simulation</span></span>
<span><span class="va">chin_sims_by_gc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                     alpha_repunit <span class="op">=</span> <span class="va">arep</span>,</span>
<span>                     return_indiv_posteriors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     resampling_unit <span class="op">=</span> <span class="st">"gene_copies"</span><span class="op">)</span></span></code></pre></div>
<p>and here we process the output and plot it:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarise things</span></span>
<span><span class="va">repu_pofzs_gc</span> <span class="op">&lt;-</span> <span class="va">chin_sims_by_gc</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit</span> <span class="op">==</span> <span class="va">simulated_repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">indiv</span>, <span class="va">simulated_collection</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>  <span class="co"># first aggregate over reporting units</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repu_PofZ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PofZ</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">repunit</span>, <span class="va">simulated_collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulated_collection <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">simulated_collection</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">simulated_collection</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'iter', 'indiv', 'simulated_collection'.</span></span>
<span><span class="co">#&gt; You can override using the `.groups` argument.</span></span>
<span></span>
<span><span class="co"># also get the number of simulated individuals from each collection</span></span>
<span><span class="va">num_simmed_gc</span> <span class="op">&lt;-</span> <span class="va">chin_sims_by_gc</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">indiv</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">simulated_collection</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># note, the last few steps make simulated collection a factor so that collections within</span></span>
<span><span class="co"># the same repunit are grouped together in the plot.</span></span>
<span></span>
<span><span class="co"># now, plot it</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">repu_pofzs_gc</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">simulated_collection</span>, y <span class="op">=</span> <span class="va">repu_PofZ</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">num_simmed_gc</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fl">1.025</span>, label <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">0</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">9</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">1.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="tools/boxplot-pofz-gc-sim-1.png"><!-- --></p>
<p>And in that, we find somewhat fewer fish that have low posteriors, but there are still some. This reminds us that with this dataset, (rather) occasionally it is possible to get individuals carrying genotypes that make it difficult to correctly assign them to reporting unit.</p>
</div>
<div class="section level2">
<h2 id="sub-specifying-collection-proportions-or-dirichlet-parameters">“sub-specifying” collection proportions or dirichlet parameters<a class="anchor" aria-label="anchor" href="#sub-specifying-collection-proportions-or-dirichlet-parameters"></a>
</h2>
<p>If you are simulating the reporting unit proportions or numbers, and want to have more control over which collections those fish are simulated from, within the reporting units, then the <code>sub_ppn</code> and <code>sub_dirichlet</code> settings are for you. These are given as column names in the <code>alpha_collection</code> data frame.</p>
<p>For example, let’s say we want to simulate reporting unit proportions as before, using <code>arep</code> from above:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arep</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   repunit                 dirichlet</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                       &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 CentralValleyfa            8.21  </span></span>
<span><span class="co">#&gt; 2 KlamathR                   0.669 </span></span>
<span><span class="co">#&gt; 3 RogueR                     0.606 </span></span>
<span><span class="co">#&gt; 4 CaliforniaCoast            0.300 </span></span>
<span><span class="co">#&gt; 5 NCaliforniaSOregonCoast    0.0923</span></span>
<span><span class="co">#&gt; 6 UColumbiaRsufa             0.0430</span></span></code></pre></div>
<p>But, now, let’s say that within reporting unit we want specific weights for different collections. Then we could specify those, for example, like this:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arep_subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">collection</span>, <span class="op">~</span><span class="va">sub_ppn</span>,</span>
<span>  <span class="st">"Eel_R"</span>,   <span class="fl">0.1</span>,</span>
<span>  <span class="st">"Russian_R"</span>, <span class="fl">0.9</span>,</span>
<span>  <span class="st">"Butte_Cr_fa"</span>, <span class="fl">0.7</span>,</span>
<span>  <span class="st">"Feather_H_sp"</span>, <span class="fl">0.3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Collections that are not listed are given equal proportions <em>within repunits that had no collections listed</em>. However, if a collection is not listed, but other collections within its repunit are, then its simulated proportion will be zero. (Technically, it is not zero, but it is so small—like <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>10</mn><mrow><mi>−</mi><mn>8</mn></mrow></msup><annotation encoding="application/x-tex">10^{-8}</annotation></semantics></math> that is is effectively 0…doing that made coding it up a lot easier…)</p>
<p>Now, we can simulate with that and see what the resulting proportion of fish from each collection is:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chin_sims_sub_ppn</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                     alpha_repunit <span class="op">=</span> <span class="va">arep</span>,</span>
<span>                     alpha_collection <span class="op">=</span> <span class="va">arep_subs</span>,</span>
<span>                     return_indiv_posteriors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># don't bother returning individual posteriors</span></span></code></pre></div>
<p>Now observe the average proportions of the collections and repunits that were simulated, and the average fraction, <em>within reporting units</em> of each of the collection</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chin_sims_sub_ppn</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">repunit</span>, <span class="va">collection</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">true_pi</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>repunit_mean_pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mean_pi</span><span class="op">)</span>,</span>
<span>         fract_within <span class="op">=</span> <span class="va">mean_pi</span> <span class="op">/</span> <span class="va">repunit_mean_pi</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>fract_within <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">fract_within</span> <span class="op">&lt;</span> <span class="fl">1e-06</span>, <span class="fl">0</span>, <span class="va">fract_within</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="co"># anything less than 1 in a million gets called 0</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit_mean_pi</span> <span class="op">&gt;</span> <span class="fl">0.0</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'repunit'. You can override using the</span></span>
<span><span class="co">#&gt; `.groups` argument.</span></span>
<span><span class="co">#&gt; # A tibble: 20 × 5</span></span>
<span><span class="co">#&gt; # Groups:   repunit [6]</span></span>
<span><span class="co">#&gt;    repunit                 collection       mean_pi repunit_mean_pi fract_within</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                   &lt;chr&gt;              &lt;dbl&gt;           &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 CaliforniaCoast         Eel_R            3.65e-3          0.0365        0.1  </span></span>
<span><span class="co">#&gt;  2 CaliforniaCoast         Russian_R        3.29e-2          0.0365        0.9  </span></span>
<span><span class="co">#&gt;  3 CentralValleyfa         Battle_Cr        8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt;  4 CentralValleyfa         Butte_Cr_fa      5.77e-1          0.825         0.700</span></span>
<span><span class="co">#&gt;  5 CentralValleyfa         Deer_Cr_fa       8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt;  6 CentralValleyfa         Feather_H_fa     8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt;  7 CentralValleyfa         Feather_H_sp     2.47e-1          0.825         0.300</span></span>
<span><span class="co">#&gt;  8 CentralValleyfa         Mill_Cr_fa       8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt;  9 CentralValleyfa         Mokelumne_R_fa   8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt; 10 CentralValleyfa         Sacramento_R_lf  8.25e-8          0.825         0    </span></span>
<span><span class="co">#&gt; 11 KlamathR                Klamath_IGH_fa   3.07e-2          0.0613        0.5  </span></span>
<span><span class="co">#&gt; 12 KlamathR                Trinity_H_sp     3.07e-2          0.0613        0.5  </span></span>
<span><span class="co">#&gt; 13 NCaliforniaSOregonCoast Chetco_R         5.77e-3          0.0115        0.5  </span></span>
<span><span class="co">#&gt; 14 NCaliforniaSOregonCoast Smith_R          5.77e-3          0.0115        0.5  </span></span>
<span><span class="co">#&gt; 15 RogueR                  Applegate_Cr     2.78e-2          0.0557        0.5  </span></span>
<span><span class="co">#&gt; 16 RogueR                  Cole_Rivers_H    2.78e-2          0.0557        0.5  </span></span>
<span><span class="co">#&gt; 17 UColumbiaRsufa          Hanford_Reach    2.59e-3          0.0104        0.25 </span></span>
<span><span class="co">#&gt; 18 UColumbiaRsufa          PriestRapids_H   2.59e-3          0.0104        0.25 </span></span>
<span><span class="co">#&gt; 19 UColumbiaRsufa          Wells_H          2.59e-3          0.0104        0.25 </span></span>
<span><span class="co">#&gt; 20 UColumbiaRsufa          Wenatchee_R      2.59e-3          0.0104        0.25</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multiple-simulation-scenarios-and-100-simulations">Multiple simulation scenarios and “100% Simulations”<a class="anchor" aria-label="anchor" href="#multiple-simulation-scenarios-and-100-simulations"></a>
</h2>
<p>In the fisheries world, “100% simulations” have been a staple. In these simulations, mixtures are simulated in which 100% of the individuals are from one collection (or reporting unit, I suppose). Eric has never been a big fan of these since they don’t necessarily tell you how you might do inferring actual mixtures that you might encounter. Nonetheless, since they have been such a mainstay in the field, it is worthwile showing how to do 100% simulations using <code>rubias</code>. Furthermore, when people asked for this feature it made it clear that Eric had to provide a way to simulate multiple different scenarios without re-processing the reference data set each time. So this is what I came up with: the way we do it is to pass a <em>list</em> of scenarios to the <code>alpha_repunit</code> or <code>alpha_collection</code> option in <code><a href="reference/assess_reference_loo.html">assess_reference_loo()</a></code>. These can be named lists, if desired. So, for example, let’s do 100% simulations for each of the repunits in <code>arep</code>:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arep</span><span class="op">$</span><span class="va">repunit</span></span>
<span><span class="co">#&gt; [1] "CentralValleyfa"         "KlamathR"               </span></span>
<span><span class="co">#&gt; [3] "RogueR"                  "CaliforniaCoast"        </span></span>
<span><span class="co">#&gt; [5] "NCaliforniaSOregonCoast" "UColumbiaRsufa"</span></span></code></pre></div>
<p>We will let the collections within them just be drawn from a dirichlet distribution with parameter 10 (so, pretty close to equal proportions).</p>
<p>So, to do this, we make a list of data frames with the proportions. We’ll give it some names too:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">six_hundy_scenarios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">arep</span><span class="op">$</span><span class="va">repunit</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>repunit <span class="op">=</span> <span class="va">x</span>, ppn <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">six_hundy_scenarios</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"All"</span>, <span class="va">arep</span><span class="op">$</span><span class="va">repunit</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we use it, producing only 5 replicates for each scenario:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">repu_hundy_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                     alpha_repunit <span class="op">=</span> <span class="va">six_hundy_scenarios</span>,</span>
<span>                     alpha_collection <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-CentralValleyfa with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-KlamathR with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-RogueR with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-CaliforniaCoast with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-NCaliforniaSOregonCoast with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="co">#&gt; ++++ Starting in on repunit_scenario All-UColumbiaRsufa with collection scenario 1 ++++</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 1 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 2 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 3 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 4 of 5</span></span>
<span><span class="co">#&gt; Doing LOO simulations rep 5 of 5</span></span>
<span><span class="va">repu_hundy_results</span></span>
<span><span class="co">#&gt; # A tibble: 2,070 × 9</span></span>
<span><span class="co">#&gt;    repunit_scenario   collection_scenario  iter repunit collection true_pi     n</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;               &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 All-CentralValley… 1                       1 Centra… Feather_H…   0.100     5</span></span>
<span><span class="co">#&gt;  2 All-CentralValley… 1                       1 Centra… Butte_Cr_…   0         0</span></span>
<span><span class="co">#&gt;  3 All-CentralValley… 1                       1 Centra… Mill_Cr_sp   0         0</span></span>
<span><span class="co">#&gt;  4 All-CentralValley… 1                       1 Centra… Deer_Cr_sp   0         0</span></span>
<span><span class="co">#&gt;  5 All-CentralValley… 1                       1 Centra… UpperSacr…   0         0</span></span>
<span><span class="co">#&gt;  6 All-CentralValley… 1                       1 Centra… Feather_H…   0.141     5</span></span>
<span><span class="co">#&gt;  7 All-CentralValley… 1                       1 Centra… Butte_Cr_…   0.100     5</span></span>
<span><span class="co">#&gt;  8 All-CentralValley… 1                       1 Centra… Mill_Cr_fa   0.140     5</span></span>
<span><span class="co">#&gt;  9 All-CentralValley… 1                       1 Centra… Deer_Cr_fa   0.193    14</span></span>
<span><span class="co">#&gt; 10 All-CentralValley… 1                       1 Centra… Mokelumne…   0.102     5</span></span>
<span><span class="co">#&gt; # ℹ 2,060 more rows</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: post_mean_pi &lt;dbl&gt;, mle_pi &lt;dbl&gt;</span></span></code></pre></div>
<div class="section level3">
<h3 id="do-it-again-with-100-collections">Do it again with 100% collections<a class="anchor" aria-label="anchor" href="#do-it-again-with-100-collections"></a>
</h3>
<p>Just to make sure that it is clear how to do this with collections (rather than reporting units) as well, lets do 100% simulations for a handful of the collections. Let’s just randomly take 5 of them, and do 6 reps for each:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">hundy_colls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">chinook</span><span class="op">$</span><span class="va">collection</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">hundy_colls</span></span>
<span><span class="co">#&gt; [1] "Deer_Cr_fa"  "Kitwanga_R"  "Morice_R"    "Wenatchee_R" "Russian_R"</span></span></code></pre></div>
<p>So, now make a list of those with 100% specifications in the tibbles:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hundy_coll_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">hundy_colls</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="va">x</span>, ppn <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"100%"</span>, <span class="va">hundy_colls</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then, do it:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hundy_coll_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assess_reference_loo.html">assess_reference_loo</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                     gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                     reps <span class="op">=</span> <span class="fl">6</span>, </span>
<span>                     mixsize <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                     alpha_collection <span class="op">=</span> <span class="va">hundy_coll_list</span><span class="op">)</span></span>
<span><span class="va">hundy_coll_results</span></span>
<span><span class="co">#&gt; # A tibble: 2,070 × 9</span></span>
<span><span class="co">#&gt;    repunit_scenario collection_scenario  iter repunit   collection true_pi     n</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;               &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 1                100%_Deer_Cr_fa         1 CentralV… Feather_H…       0     0</span></span>
<span><span class="co">#&gt;  2 1                100%_Deer_Cr_fa         1 CentralV… Butte_Cr_…       0     0</span></span>
<span><span class="co">#&gt;  3 1                100%_Deer_Cr_fa         1 CentralV… Mill_Cr_sp       0     0</span></span>
<span><span class="co">#&gt;  4 1                100%_Deer_Cr_fa         1 CentralV… Deer_Cr_sp       0     0</span></span>
<span><span class="co">#&gt;  5 1                100%_Deer_Cr_fa         1 CentralV… UpperSacr…       0     0</span></span>
<span><span class="co">#&gt;  6 1                100%_Deer_Cr_fa         1 CentralV… Feather_H…       0     0</span></span>
<span><span class="co">#&gt;  7 1                100%_Deer_Cr_fa         1 CentralV… Butte_Cr_…       0     0</span></span>
<span><span class="co">#&gt;  8 1                100%_Deer_Cr_fa         1 CentralV… Mill_Cr_fa       0     0</span></span>
<span><span class="co">#&gt;  9 1                100%_Deer_Cr_fa         1 CentralV… Deer_Cr_fa       1    50</span></span>
<span><span class="co">#&gt; 10 1                100%_Deer_Cr_fa         1 CentralV… Mokelumne…       0     0</span></span>
<span><span class="co">#&gt; # ℹ 2,060 more rows</span></span>
<span><span class="co">#&gt; # ℹ 2 more variables: post_mean_pi &lt;dbl&gt;, mle_pi &lt;dbl&gt;</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="bootstrap-corrected-reporting-unit-proportions">Bootstrap-Corrected Reporting Unit Proportions<a class="anchor" aria-label="anchor" href="#bootstrap-corrected-reporting-unit-proportions"></a>
</h1>
<p>These are obtained using <code>method = "PB"</code> in <code><a href="reference/infer_mixture.html">infer_mixture()</a></code>. When invoked, this will return the regular MCMC results as before, but also will population the <code>bootstrapped_proportions</code> field of the output. Doing so takes a little bit longer, computationally, because there is a good deal of simulation involved:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix_est_pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">chinook</span>, </span>
<span>                         mixture <span class="op">=</span> <span class="va">chinook_mix</span>, </span>
<span>                         gen_start_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                         method <span class="op">=</span> <span class="st">"PB"</span><span class="op">)</span></span></code></pre></div>
<p>And now we can compare the estimates, showing here the 10 most prevalent repunits, in the <code>rec1</code> fishery:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix_est_pb</span><span class="op">$</span><span class="va">mixing_proportions</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repprop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">mix_est_pb</span><span class="op">$</span><span class="va">bootstrapped_proportions</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">mixture_collection</span> <span class="op">==</span> <span class="st">"rec1"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">repprop</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>It gives us a result that we expect: no appreciable difference, because the reporting units are already very well resolved, so we don’t expect that the parametric bootstrap procedure would find any benefit in correcting them.</p>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Andersonetal2008" class="csl-entry">
Anderson, Eric C, Robin S Waples, and Steven T Kalinowski. 2008. “An Improved Method for Predicting the Accuracy of Genetic Stock Identification.” <em>Can J Fish Aquat Sci</em> 65: 1475–86.
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=rubias" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/eriqande/rubias/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/eriqande/rubias/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>CC0</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing rubias</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Eric C. Anderson <br><small class="roles"> Author, maintainer </small>   </li>
<li>Ben Moran <br><small class="roles"> Author </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://CRAN.R-project.org/package=rubias" class="external-link"><img src="https://www.r-pkg.org/badges/version/rubias" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric C. Anderson, Ben Moran.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
