<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Summarizing Uncertainty on Stock-Specific Total Catch • rubias</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing Uncertainty on Stock-Specific Total Catch">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rubias</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/rubias-fully-bayesian.html">Using the Fully Bayesian Model in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-overview.html">An Overview of rubias Usage</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-underlying-data-structures.html">An Explanation of the Underlying Data Structures in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/uncertainty-on-total-catch.html">Summarizing Uncertainty on Stock-Specific Total Catch</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/eriqande/rubias/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Summarizing Uncertainty on Stock-Specific Total Catch</h1>
                        <h4 data-toc-skip class="author">Eric C.
Anderson</h4>
            
            <h4 data-toc-skip class="date">2025-10-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/eriqande/rubias/blob/HEAD/vignettes/uncertainty-on-total-catch.Rmd" class="external-link"><code>vignettes/uncertainty-on-total-catch.Rmd</code></a></small>
      <div class="d-none name"><code>uncertainty-on-total-catch.Rmd</code></div>
    </div>

    
    
<p>The R package ‘rubias,’ like most other mixed-stock analysis
programs, implements a typical finite mixture model. In such a model, a
key parameter is the proportion of different components present in the
population from which the samples are being taken. We call this
parameter
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝛑</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>π</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>π</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\boldsymbol{\pi}= (\pi_1,\ldots,\pi_K)</annotation></semantics></math>
which includes a proportion for each of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
components in the mixture. In the mixed-stock fisheries context,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
holds the stock-specific proportions that are present <em>in the
ocean</em> at the time and location of sampling. This assemblage of fish
is treated in the model (because of the assumption of multinomial
sampling) as an infinite pool from which samples are drawn.</p>
<p>While
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
is clearly a necessary variable in a mixture model—it is essential to
model the population proportions—it might not always be the ultimate
target of interest. A germane example of this occurs when directly
managing fisheries using genetic stock identification (GSI). In these
cases, managers may be particularly interested in the numbers of fish
from different populations or reporting units that occur in the total
catch from a particular fishery (or across multiple fisheries).</p>
<p>Because fisheries might be managed to limit impacts on a particular
weak stock to below a certain threshold, it is important to be able to
evaluate the uncertainty in these estimates of stock-specific total
catch. This vignette describes how to do that with the new functionality
available in the updated version of ‘rubias’. Before describing the
details, however, we explain why propagating the uncertainty found in
the Monte Carlo sample of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
does not accurately reflect uncertainty in stock-specific total catch,
and we describe the newly implemented approach in ‘rubias’ for
evaluating uncertainty in stock-specific total catch in a rigorous
way.</p>
<div class="section level2">
<h2 id="uncertainty-in-boldsymbolpi-is-not-uncertainty-in-stock-specific-total-catch">Uncertainty in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
is not uncertainty in stock-specific total catch<a class="anchor" aria-label="anchor" href="#uncertainty-in-boldsymbolpi-is-not-uncertainty-in-stock-specific-total-catch"></a>
</h2>
<p>Although the uncertainty in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
can affect the amount of uncertainty in stock-specific total catch, the
two cannot be directly equated because uncertainty in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
always reflects the variability due to having drawn a random sample. By
contrast, the sample itself might be a fixed part of the total catch.
The implications of this are easiest to understand when the entire catch
has been sampled for GSI. For, example, imagine that the total catch was
300 fish, all 300 of those fish are subject to GSI, and the population
of origin of fish from stock
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
can be identified with no uncertainty using genetics. If 20 fish
originated from stock
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>,
then the total catch from stock
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
is 20, with no error. However, the 90% credible interval on the
proportion of fish from stock
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
in the ocean would be, from the binomial distribution, (0.043, 0.09). If
this variation were applied to the total catch of 300, it would
erroneously suggest that between 13 and 27 fish from stock
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
were in the catch, when it is clear that exactly 20 fish from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
were in the catch!</p>
<p>The above example is simple and contrived but it illustrates that the
uncertainty around
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
cannot be directly used to quantify the uncertainty in stock-specific
total catch. Unlike our simple example, stock-specific total catch will
typically be affected by other sources of variation such as uncertainty
in genetic assignment, subsampling of the total catch for genetic
analysis, and simple uncertainty in the estimates of the actual total
catch. Fortunately, all of these sources of uncertainty can now be
accounted for in a rigorous way within the Markov Chain Monte Carlo
(MCMC) sampling framework of ‘rubias,’ providing users with an MCMC
sample of the stock-specific total catch for each stock. This MCMC
sample can then be summarized however desired, or it can be fed into
further analyses. The next section describes the statistical background
on how ‘rubias’ provides an MCMC sample of the stock-specific total
catch.</p>
</div>
<div class="section level2">
<h2 id="mcmc-for-stock-specific-total-catch">MCMC for stock-specific total catch<a class="anchor" aria-label="anchor" href="#mcmc-for-stock-specific-total-catch"></a>
</h2>
<div class="figure" style="text-align: center">
<img src="images/standard-model-crop.png" alt="Acyclic directed graph showing the standard conditional GSI model. Diamond nodes are deterministic values used for priors. Circular nodes are variables. The shaded, circular nodes denote variables that are observed and the unshaded circular nodes are variables whose values are updated and sampled during the MCMC process. The large rectangles are plates that represent replication over individuals and/or loci." width="30%"><p class="caption">
Acyclic directed graph showing the standard conditional GSI model.
Diamond nodes are deterministic values used for priors. Circular nodes
are variables. The shaded, circular nodes denote variables that are
observed and the unshaded circular nodes are variables whose values are
updated and sampled during the MCMC process. The large rectangles are
plates that represent replication over individuals and/or loci.
</p>
</div>
<p>The above figure shows the acyclic directed graph (DAG) used in the
conditional GSI model implemented in ‘rubias.’ The stock proportion
vector in the ocean,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>,
is visible immediately below
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>𝛌</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>π</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><annotation encoding="application/x-tex">\boldsymbol{\lambda}^{(\pi)}</annotation></semantics></math>,
which is a parameter vector for the prior on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>.
There are
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>M</mi><annotation encoding="application/x-tex">M</annotation></semantics></math>
fish in the reference baseline data set.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>𝐙</mi><mi>j</mi><mo>*</mo></msubsup><annotation encoding="application/x-tex">\boldsymbol{Z}^*_j</annotation></semantics></math>
is the known origin of fish
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
in the baseline, while
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>𝐘</mi><mi>j</mi><mo>*</mo></msubsup><annotation encoding="application/x-tex">\boldsymbol{Y}^*_j</annotation></semantics></math>
is its observed genotype.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐙</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{Z}_i</annotation></semantics></math>
is an indicator vector giving the unknown origin of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>i</mi><mrow><mi mathvariant="normal">t</mi><mi mathvariant="normal">h</mi></mrow></msup><annotation encoding="application/x-tex">i^\mathrm{th}</annotation></semantics></math>
fish from the mixture, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐘</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{Y}_i</annotation></semantics></math>
is the genotype data from that fish,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝛉</mi><mo>ℓ</mo></msub><annotation encoding="application/x-tex">\boldsymbol{\theta}_\ell</annotation></semantics></math>
are the unknown population allele freqeunces at locus
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>ℓ</mo><annotation encoding="application/x-tex">\ell</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝛄</mi><mo>ℓ</mo></msub><annotation encoding="application/x-tex">\boldsymbol{\gamma}_\ell</annotation></semantics></math>
are the priors on those frequencies. MCMC sampling in this model
proceeds using Gibbs sampling, with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐙</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{Z}_i</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝛉</mi><mo>ℓ</mo></msub><annotation encoding="application/x-tex">\boldsymbol{\theta}_\ell</annotation></semantics></math>
being updated during each sweep of the algorithm by drawing each new
value from its full conditional distribution. (A “sweep” in this context
is a cycle of the MCMC algorithm during which values have been updated
for all the variables in the model.)</p>
<div class="figure" style="text-align: center">
<img src="images/stock-spec-catch-model-crop.png" alt="Acyclic directed graph showing the augmented standard conditional GSI model to include additional fish in the catch that were not included in the GSI analysis." width="42%"><p class="caption">
Acyclic directed graph showing the augmented standard conditional GSI
model to include additional fish in the catch that were not included in
the GSI analysis.
</p>
</div>
<p>In order to sample from a posterior distribution for the
stock-specific total catch, we introduce the indicator vectors
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup><annotation encoding="application/x-tex">\boldsymbol{Z}^r_k</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>R</mi></mrow><annotation encoding="application/x-tex">k=1,\ldots,R</annotation></semantics></math>,
giving the unknown origin of each of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
fish in the total catch that remain after the fish used in the GSI
analysis are removed from the total catch. The DAG for that model
appears in the above figure. Sampling in this new model proceeds exactly
as before, except that in every sweep we also simulate values for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
different
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup><annotation encoding="application/x-tex">\boldsymbol{Z}^r_k</annotation></semantics></math>
vectors from their full conditional distribution, which is simply a
multinomial sample of size 1 from cell probabilities that are the
current values of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup><mo>∼</mo><mrow><mi mathvariant="normal">M</mi><mi mathvariant="normal">u</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">t</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">n</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">m</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>,</mo><mi>𝛑</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mspace width="0.222em"></mspace><mspace width="0.222em"></mspace><mrow><mi mathvariant="normal">f</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">r</mi></mrow><mspace width="0.222em"></mspace><mi>k</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>R</mi><mi>.</mi></mrow><annotation encoding="application/x-tex">
\boldsymbol{Z}^r_k \sim \mathrm{Multinomial}(1, \boldsymbol{\pi}),~~\mathrm{for}~k=1,\ldots,R.
</annotation></semantics></math> At the conclusion of each sweep a
sample from the posterior distribution of the stock-specific total catch
is obtained by summing over all the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup><annotation encoding="application/x-tex">\boldsymbol{Z}^r_k</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐙</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{Z}_i</annotation></semantics></math>
vectors,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mrow><mi mathvariant="normal">s</mi><mi mathvariant="normal">s</mi></mrow></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>R</mi></munderover><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup><mo>+</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>N</mi></munderover><msub><mi>𝐙</mi><mi>i</mi></msub><mi>.</mi></mrow><annotation encoding="application/x-tex">
C_\mathrm{ss} = \sum_{k=1}^R \boldsymbol{Z}^r_k + \sum_{i=1}^N \boldsymbol{Z}_i.
</annotation></semantics></math> In fact, to make things even easier, by
definition
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>R</mi></msubsup><msubsup><mi>𝐙</mi><mi>k</mi><mi>r</mi></msubsup></mrow><annotation encoding="application/x-tex">\sum_{k=1}^R \boldsymbol{Z}^r_k</annotation></semantics></math>
is simply a multinomial random variable of R trials with cell
probabilities of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>.</p>
<div class="section level3">
<h3 id="additional-complications">Additional Complications<a class="anchor" aria-label="anchor" href="#additional-complications"></a>
</h3>
<p>Sometimes, the size of the total catch,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>,
might itself be estimated with some uncertainty. Although catch
estimates are often presented without uncertainty, validated catch
estimates may be rare. Frequently, catches of individuals are estimated
by converting landed biomass into individual pieces by estimating
average weight for countable subsamples of the catch (e.g. commercial
offloads at fish processing facilities), and substantial uncertainty in
catch numbers may result due to heterogeneity in average size by date,
location, or fishing gear, or even due to error in handled samples
[e.g., species identification for estimating average weights; see <span class="citation">Faunce et al. (2015)</span>]. Another method commonly
used to estimate numerical catch is to multiply total fishing effort by
sampled estimates of catch per unit of effort (CPUE, e.g. creel surveys
in recreational fisheries). Because fishers can vary strongly based on
their behavior (spatiotemporal distribution, skill, and willingness to
be surveyed), these estimates based on CPUE can be highly uncertain
<span class="citation">(Cabanellas-Reboredo et al. 2017)</span>.</p>
<p>In order to handle uncertainty in the true total catch, in ‘rubias’
the assumption is made that the true size of the total catch is
conditionally independent of the stock composition given the estimated
total catch. With that assumption, the uncertainty in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
(the total catch) can be propagated into the stock-specific total catch
by supplying a sample of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
from its posterior distribution. Different values from that posterior
sample can be used for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
in each sweep that ‘rubias’ performs. Examples for doing so are provided
below. It is worth noting that there may be times when the stock
composition is not conditionally independent of the true catch given the
estimated total catch, and users should verify whether or not this
assumption holds. For example, one may imagine a case in which
transgenerational genetic mark-recapture <span class="citation">(Rosenbaum et al. 2024)</span> would be used to inform
run reconstructions and the associated catch estimates for some subset
of stocks, and GSI could be used to parse the total catch estimates into
all stocks, resulting in genotypes possibly contributing to both total
catch estimates and estimates of stock composition.</p>
<p>Often, the mixed fishery includes commercial or recreational
landings, which are certainly part of the lethal catch or take. However,
in some mixed stock fisheries, some of the fish sampled and genotyped
might be from a test fishery which is not necessarily lethal, as fish
might be returned alive to the water after being sampled for genetic
analysis. In these cases, fish from the test fishery might not be
considered part of the total catch, and therefore, they should not
contribute to the estimated impacts on any particular stock that they
are assigned to. In other cases, managers might have an independent
estimate of post-sampling mortality that is not 0. This can be
accommodated, as shown in the examples in the “Some Genotypes Sampled
Might not be Considered Part of the Catch/Mortality” section below.</p>
</div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>In this section we walk through some simple, hands-on examples. We
will be using some of the example data that comes with the rubias
package. Namely, we are using the <code>chinook</code> reference data
set that comprises 91 SNP markers typed on 7,301 fish from 69
collections belonging to 39 different reporting units.</p>
<p>The samples from the mixed stock fisheries we will analyze are in the
data object <code>chinook_mix</code>. This data set contains genotypes
from 2,256 fish grouped into three different fisheries (“rec1”, “rec2”,
and “rec3”). (These were not really different fisheries, but for our
purposes, here, we are going to pretend that they are different
fisheries featuring different levels of management information.) For the
most part, the mix of fish found in these fisheries are typical of those
seen off the coast of central and northern California. Most of the fish
are from the Central Valley fall-run stock and, for our purposes here,
we will assume that we are particularly interested in impacts on one of
the less numerous reporting units in these mixtures: “CaliforniaCoast”,
which comprises fish from the Eel and Russian rivers,listed as
Threatened under the U.S. Endanged Species Act.</p>
<p>For the purposes of illustration, we will imagine that “rec1” is a
recreational fishery in which every single landed fish is sampled,
“rec2” is a fishery in which an exact 25% of the total catch is sampled
and genotyped, and “rec3” is a fishery from which the managers try to
sample 25% of the fish; however the true fraction sampled varies quite a
bit. Let’s imagine that the managers of “rec3” have a Bayesian method to
estimate the true fraction sampled, and that the posterior for that
fraction is a beta distribution with parameters 2 and 6, such that a
sample of that posterior can be obtained like this</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eriqande/rubias" class="external-link">rubias</a></span><span class="op">)</span></span>
<span><span class="va">rubi_method</span> <span class="op">&lt;-</span>  <span class="st">"MCMC"</span> <span class="co"># set this to BR if you want to run everything using</span></span>
<span>                       <span class="co"># the "Bayesian Resampling" approach in which the allele </span></span>
<span>                       <span class="co"># frequencies are updated according to the allocations</span></span>
<span>                       <span class="co"># of the fish in the mixture.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">rec3_tot_catch_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">16</span>, <span class="fl">48</span><span class="op">)</span></span></code></pre></div>
<p>And that posterior sample is distributed as shown in the following
figure.</p>
<div class="figure" style="text-align: center">
<img src="uncertainty-on-total-catch_files/figure-html/rec3hist-1.png" alt="Histogram of the estimated sampling fraction for the rec3 fishery" width="70%"><p class="caption">
Histogram of the estimated sampling fraction for the rec3 fishery
</p>
</div>
<p>In other words, there is a lot of uncertainty about what the sampling
fraction is, which means there is a lot of uncertainty in the estimate
of total catch from that “rec3” fishery.</p>
<div class="section level3">
<h3 id="using-variability-in-boldsymbolpi">Using Variability in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math><a class="anchor" aria-label="anchor" href="#using-variability-in-boldsymbolpi"></a>
</h3>
<p>We will start off with a simple look at how things can go wrong if
you use the variability in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>
to try to estimate the variability on total catch. We will focus on the
“rec1” fishery here, in which we are assuming that every single fish in
the catch has been sampled and genotyped.</p>
<p>First, we analyze just that fishery and summarise by reporting
units.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rec1_ests</span> <span class="op">&lt;-</span> <span class="va">chinook_mix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">collection</span> <span class="op">==</span> <span class="st">"rec1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span><span class="va">chinook</span>, <span class="va">.</span>, gen_start_col <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Collating data; compiling reference allele frequencies, etc.   time: 1.01 seconds</span></span>
<span><span class="co">## Computing reference locus specific means and variances for computing mixture z-scores   time: 0.10 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec1 with 743 individuals</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.52 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.03 seconds</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the mixing proportions trace by repunit</span></span>
<span><span class="va">rec1_pi_trace_repu</span> <span class="op">&lt;-</span> <span class="va">rec1_ests</span><span class="op">$</span><span class="va">mix_prop_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'sweep'. You can override using the</span></span>
<span><span class="co">## `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the maximum-a-posteriori assignment to repunit</span></span>
<span><span class="va">rec1_MAP_repu</span> <span class="op">&lt;-</span> <span class="va">rec1_ests</span><span class="op">$</span><span class="va">indiv_posteriors</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">indiv</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_PofZ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PofZ</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">indiv</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">repunit_PofZ</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">repunit_PofZ</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'indiv'. You can override using the</span></span>
<span><span class="co">## `.groups` argument.</span></span></code></pre>
When we look at the <em>maximum-a-posteriori</em> (MAP) assignments of
individuals to reporting unit, we find 22 fish assigned to the
CaliforniaCoast, most of them with a posterior probability near 1.0. The
histogram of their posterior probabilities appears in the figure below.
<div class="figure" style="text-align: center">
<img src="uncertainty-on-total-catch_files/figure-html/posthist-1.png" alt="Histogram of posterior probabilites of membership in the CaliforniaCoast reporting unit for 22 fish assigned there." width="70%"><p class="caption">
Histogram of posterior probabilites of membership in the CaliforniaCoast
reporting unit for 22 fish assigned there.
</p>
</div>
<p>It appears that 15 fish are almost certainly from CaliforniaCoast,
with another 5 being very likely from there, and then two of them
somewhat likely from there. In other words, we are pretty confident that
20 to 22 fish from the CaliforniaCoast are included in the total
catch.</p>
Let us see what happens if we naively use the estimated proportion of
fish from CaliforniaCoast to try to estimate the number of such fish in
the total catch. The figure below is a histogram of the posterior sample
of the fraction of fish from CaliforniaCoast in the ocean from which the
catch was taken:
<div class="figure" style="text-align: center">
<img src="uncertainty-on-total-catch_files/figure-html/pitraces-1.png" alt="Histogram of a sample from the posterior distribution of the fraction of fish from the CaliforniaCoast reporting unit in the ocean." width="70%"><p class="caption">
Histogram of a sample from the posterior distribution of the fraction of
fish from the CaliforniaCoast reporting unit in the ocean.
</p>
</div>
If we used that fraction to try to estimate the uncertainty around the
total catch of CaliforniaCoast we would greatly overestimate that
uncertainty. By multiplying the fraction of catch by the total number of
fish (which happens to be 743 fish), we would get the posterior for the
total catch of CaliforniaCoast as shown in the following figure.
<div class="figure" style="text-align: center">
<img src="uncertainty-on-total-catch_files/figure-html/bogusfract-1.png" alt="Posterior distribution of total catch from CaliforniaCoast if calculated naïvely by simply multiplying the corresponding mixing proportion sample by the total number of fish caught." width="70%"><p class="caption">
Posterior distribution of total catch from CaliforniaCoast if calculated
naïvely by simply multiplying the corresponding mixing proportion sample
by the total number of fish caught.
</p>
</div>
<p>This is absurd! On the basis of individual assignments, we are pretty
sure that between 20 and 22 CaliforniaCoast fish were caught, certainly
not 14 or 30! This is why it is important to use the new functionality
in ‘rubias’ to estimate total stock-specific catch. We illustrate that
in the next two sections.</p>
</div>
<div class="section level3">
<h3 id="all-fishery-samples-are-part-of-the-catch">All Fishery Samples are Part of the Catch<a class="anchor" aria-label="anchor" href="#all-fishery-samples-are-part-of-the-catch"></a>
</h3>
<p>We start off with what will likely be the most standard case: the
samples have come from the fishery and all the samples are considered to
be part of the impact of the fishery. In other words, every fish in the
sample of genotypes is considered to be part of the total catch. (The
next section discusses how to handle departures from this standard
case).</p>
<p>To estimate the total stock specific catch we must provide ‘rubias’
with information about the size of the total catch. We provide this
information in a tibble with one column <code>collection</code> which
gives the mixture collection (“rec1”, “rec2”, and “rec3” in the current
case), and another column, <code>tot_catch</code>, <em>which must be a
list-column</em>, that gives the information about the total catch. This
second column has to be a list column because we want it to be able to
hold a sample from the posterior distribution for total, aggregate
catch.</p>
<p>In order to make this tibble, we will need to know the sample sizes
from our three fisheries. They are:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chinook_mix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">collection</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span></span>
<span><span class="co">##   collection     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> rec1         743</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> rec2         772</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> rec3         741</span></span></code></pre>
<p>So, given what we described earlier about how the sample size relates
to the total catch (and, specifically recalling that
<code>rec3_tot_catch_sample</code> is a sample from the posterior
distribution of sampled catch fractions), we form the necessary tibble
like this:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_catches</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  collection <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rec1"</span>, <span class="st">"rec2"</span>, <span class="st">"rec3"</span><span class="op">)</span>,</span>
<span>  tot_catch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fl">743</span>,</span>
<span>    <span class="fl">772</span> <span class="op">/</span> <span class="fl">0.25</span>,</span>
<span>    <span class="fl">743</span> <span class="op">/</span> <span class="va">rec3_tot_catch_sample</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>It looks like this:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_catches</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span></span>
<span><span class="co">##   collection tot_catch    </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> rec1       <span style="color: #949494;">&lt;dbl [1]&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> rec2       <span style="color: #949494;">&lt;dbl [1]&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> rec3       <span style="color: #949494;">&lt;dbl [5,000]&gt;</span></span></span></code></pre>
<p>Then, to get these standard-case total catch estimates from ‘rubias’
we just need to pass this tibble into <code><a href="../reference/infer_mixture.html">infer_mixture()</a></code> as
the <code>total_catch_tib</code> parameter.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix_ests_with_catch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">chinook</span>,</span>
<span>  mixture <span class="op">=</span> <span class="va">chinook_mix</span>,</span>
<span>  gen_start_col <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  total_catch_tib <span class="op">=</span> <span class="va">total_catches</span>,</span>
<span>  method <span class="op">=</span> <span class="va">rubi_method</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Collating data; compiling reference allele frequencies, etc.   time: 0.80 seconds</span></span>
<span><span class="co">## Computing reference locus specific means and variances for computing mixture z-scores   time: 0.10 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec2 with 772 individuals</span></span>
<span><span class="co">## tot_catch_vec: 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.58 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.06 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec1 with 743 individuals</span></span>
<span><span class="co">## tot_catch_vec: 743 743 743 743 743 743 743 743 743 743 743 743...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.55 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.06 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec3 with 741 individuals</span></span>
<span><span class="co">## tot_catch_vec: 3839 3198 2791 4069 2834 2951 2911 2258 4157 2172 3606 4042...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.57 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.06 seconds</span></span></code></pre>
<p>After that has run, the tibble in
<code>mix_ests_with_catch$stock_specific_total_catch_traces</code> holds
the posterior sample of the total catch. It is structured much like the
mixing proportions trace output. The first few lines look like:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix_ests_with_catch</span><span class="op">$</span><span class="va">stock_specific_total_catch_traces</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 414,000 × 5</span></span></span>
<span><span class="co">##    mixture_collection sweep repunit         collection            SSTC</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> rec2                   0 CentralValleyfa Feather_H_sp           256</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> rec2                   0 CentralValleysp Butte_Cr_Sp             12</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> rec2                   0 CentralValleysp Mill_Cr_sp              71</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> rec2                   0 CentralValleysp Deer_Cr_sp              70</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> rec2                   0 CentralValleysp UpperSacramento_R_sp    83</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> rec2                   0 CentralValleyfa Feather_H_fa           323</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> rec2                   0 CentralValleyfa Butte_Cr_fa            399</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> rec2                   0 CentralValleyfa Mill_Cr_fa             250</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> rec2                   0 CentralValleyfa Deer_Cr_fa             316</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> rec2                   0 CentralValleyfa Mokelumne_R_fa         322</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 413,990 more rows</span></span></span></code></pre>
<p>The column <code>SSTC</code> holds the stock-specific total
catch.</p>
<p>Thus, if we want to see what the posterior distribution for the total
catch from the CaliforniaCoast repunit looks like in each of the three
fisheries (“rec1”, “rec2”, and “rec3”) we just have to:</p>
<ol style="list-style-type: decimal">
<li>Summarize by reporting unit</li>
<li>Filter down to just CaliforniaCoast</li>
<li>Discard the burn-in (which we take to be the default 100 sweeps,
here)</li>
<li>Plot it</li>
</ol>
<p>Those steps look like this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_sstc</span> <span class="op">&lt;-</span> <span class="va">mix_ests_with_catch</span><span class="op">$</span><span class="va">stock_specific_total_catch_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_sstc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">SSTC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">CC_sstc</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">repunit_sstc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">mixture_collection</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="uncertainty-on-total-catch_files/figure-html/steps-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>That is just how we expect things to look:</p>
<ul>
<li>If we have genotyped the entire catch (as in the “rec1” fishery)
then there is very little variation around the stock specific total
catch, especially for a repunit like CaliforniaCoast that is easily
distinguished from the others.</li>
<li>If we have only genotyped 25% of the catch, there is considerably
more variation in the posterior for stock-specific total catch. This
makes sense—in this case we are extrapolating the repunit from the
mixture proportion
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛑</mi><annotation encoding="application/x-tex">\boldsymbol{\pi}</annotation></semantics></math>)
estimates to some
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>772</mn><mo>×</mo><mn>3</mn><mo>=</mo><mn>2316</mn></mrow><annotation encoding="application/x-tex">772 \times 3 = 2316</annotation></semantics></math>
fish. There will of course be a lot more uncertainty in the stock of
origin for fish that were not sampled.</li>
<li>In the “rec3” fishery, the variability is even greater because there
is uncertainty in the actual fraction of the fishery that was
sampled.</li>
</ul>
<p>It can also be instructive to visualize the total stock-specific
catch when it is decomposed into its components: 1) the fish that are
actually in the genetic sample, and 2) the additional fish (if any) that
are part of the catch but not the genetic sample. This is fairly
easy—the return object has not just
<code>stock_specific_total_catch_traces</code>, but also:</p>
<ul>
<li>
<code>allocation_count_traces</code>: the counts of fish in the
mixture sample that are allocated to the catch each iteration.</li>
<li>
<code>posterior_predictive_remaining_catch_traces</code>: the counts
of fish sampled from the posterior predictive distribution to account
for the total catch (caught fish not in the sample).</li>
</ul>
<p>We can pick those out and plot them with the total catch to confirm
things are working the way that they should.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_pprc</span> <span class="op">&lt;-</span> <span class="va">mix_ests_with_catch</span><span class="op">$</span><span class="va">posterior_predictive_remaining_catch_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_pprc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PPRC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>   <span class="co"># column PPRC has the posterior_predictive_remaining_catch</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_ac</span> <span class="op">&lt;-</span> <span class="va">mix_ests_with_catch</span><span class="op">$</span><span class="va">allocation_count_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_ac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">CA</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>  <span class="co"># column CA has the allocation_count</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># then join those to the total stock-specific catch and pivot them</span></span>
<span><span class="va">three_counts</span> <span class="op">&lt;-</span> <span class="va">CC_sstc</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">CC_pprc</span>, by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html" class="external-link">join_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">CC_ac</span>, by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html" class="external-link">join_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">repunit_sstc</span><span class="op">:</span><span class="va">repunit_ac</span>, names_to <span class="op">=</span> <span class="st">"count_type"</span>, values_to <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p>We can plot those all together like this:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">three_counts</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n</span>, fill <span class="op">=</span> <span class="va">count_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">count_type</span> <span class="op">~</span> <span class="va">mixture_collection</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Number of individuals of repunit CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<p><img src="uncertainty-on-total-catch_files/figure-html/confirm-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>It is worth taking the time to confirm for yourself that these
results are exactly as expected, given how the scenarios were
designed.</p>
</div>
<div class="section level3">
<h3 id="some-genotypes-sampled-might-not-be-considered-part-of-the-catchmortality">Some Genotypes Sampled Might not be Considered Part of the
Catch/Mortality<a class="anchor" aria-label="anchor" href="#some-genotypes-sampled-might-not-be-considered-part-of-the-catchmortality"></a>
</h3>
<p>Another possible use case could involve a test or catch-and-release
fishery that is concomitant with a retention fishery. Genotypes from
these different fisheries might all end up together as a sample from the
mixed fishery at a certain time and place. If it occurs that fish are
released after sampling and if they survive the sampling process with no
adverse impacts, one would not want to count them as part of the fishing
mortality It is, however, conceivable that there is some mortality
associated with live release, and in that case it might be appropriate
to account for these mortalities as part of the catch. A fixed rate of
mortality may be assumed and applied to all releases depending on
fishery or environmental conditions. Alternatively, when tissues are
taken from a test fishery, a score may be assigned to each fish rating
the injuries they received—and hence the chance that they might die
after release.</p>
<p>Rubias has a simple way of handling all these cases. Essentially,
each fish in the mixture is assigned a value between 0 and 1, inclusive,
in an additional column (which must appear before the start of the
genotype columns) called <code>prob_is_catch</code>. That is short for
“the probability this fish is considered to be part of the catch.” A
value of 1 means that the fish is part of the catch. A value of 0 means
that the fish is certainly <em>not</em> to be considered as part of the
catch. And a value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
that is between 0 and 1 means that the fish had a probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
of dying after sampling, and if it did die, then it should be considered
part of the catch.</p>
<p>In the previous section, <strong>All Fishery Samples are Part of the
Catch</strong>, we saw the simple case where each fish in the mixture
has <code>prob_is_catch</code> = 1. That is the default situation:
unless otherwise specified, every fish in a mixture is considered to be
part of the catch. However, if you supply the option
<code>variable_prob_is_catch = TRUE</code> to the
<code><a href="../reference/infer_mixture.html">infer_mixture()</a></code> function, then that function will require
that there is a <code>prob_is_catch</code> column, and it will use the
values within it to account for fish from test samples that might not be
considered part of the catch.</p>
<p>We will make up some data here to illustrate different scenarios. We
will imagine that:</p>
<ul>
<li>Of the 743 “rec1” fishery samples, the first 300 of them were from a
test fishery that is believed to have zero mortality and should not be
considered part of the catch.</li>
<li>All the 772 “rec2” fishery samples were taken in a test fishery
where it is well known that fish are handled roughly. On average, 1/5 of
those released have a 50% chance of dying as a result of that treatment,
and 1/5 of the individuals have a 40, 30, 20, and 10% chance of dying
and being part of the catch, respectively, as well (this is obviously
just for illustration!)</li>
<li>All the “rec3” fish are considered to be part of the catch.</li>
</ul>
<p>We can modify the chinook_mix data set to have a
<code>prob_is_catch</code> column that reflects this:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chinook_mix_with_probs</span> <span class="op">&lt;-</span> <span class="va">chinook_mix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">collection</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    prob_is_catch <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>      <span class="va">collection</span> <span class="op">==</span> <span class="st">"rec1"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">300</span> <span class="op">~</span> <span class="fl">0</span>,</span>
<span>      <span class="va">collection</span> <span class="op">==</span> <span class="st">"rec1"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">300</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>      <span class="va">collection</span> <span class="op">==</span> <span class="st">"rec2"</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      <span class="va">collection</span> <span class="op">==</span> <span class="st">"rec3"</span> <span class="op">~</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    .after <span class="op">=</span> <span class="va">indiv</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Let’s just verify that this was done correctly by counting up the
occurrences of different <code>prob_is_catch</code> values:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chinook_mix_with_probs</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">collection</span>, <span class="va">prob_is_catch</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 3</span></span></span>
<span><span class="co">##   collection prob_is_catch     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> rec1                 0     300</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> rec1                 1     443</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> rec2                 0.1   158</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> rec2                 0.2   149</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> rec2                 0.3   148</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> rec2                 0.4   160</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> rec2                 0.5   157</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> rec3                 1     741</span></span></code></pre>
<p>That looks pretty good!</p>
<p>Before we pass that off to infer_mixture() we need to address a
constraint associated with <code>variable_prob_is_catch = TRUE</code>:
the reference data set and the mixture data set must all have the same
columns. So we need to add the same column to the reference data set
<code>chinook</code>. We will simply put a column of 0’s in there.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_reference</span> <span class="op">&lt;-</span> <span class="va">chinook</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prob_is_catch <span class="op">=</span> <span class="fl">0.0</span>, .after <span class="op">=</span> <span class="va">indiv</span><span class="op">)</span></span></code></pre></div>
<p>Now, the genotype data start in the 6th column. If we run this
without the <code>variable_prob_is_catch = TRUE</code> option, then we
will get just what we got before, because <code><a href="../reference/infer_mixture.html">infer_mixture()</a></code>
will not use the <code>prob_is_catch</code> column. However, when we do
include the <code>variable_prob_is_catch = TRUE</code> option, then the
<code><a href="../reference/infer_mixture.html">infer_mixture()</a></code> function is going to look for and use the
information in the <code>prob_is_catch</code> column.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">catch_ests_with_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_mixture.html">infer_mixture</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">new_reference</span>,</span>
<span>  mixture <span class="op">=</span> <span class="va">chinook_mix_with_probs</span>,</span>
<span>  gen_start_col <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  method <span class="op">=</span> <span class="va">rubi_method</span>,</span>
<span>  total_catch_tib <span class="op">=</span> <span class="va">total_catches</span>,</span>
<span>  variable_prob_is_catch <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Collating data; compiling reference allele frequencies, etc.   time: 0.71 seconds</span></span>
<span><span class="co">## Computing reference locus specific means and variances for computing mixture z-scores   time: 0.10 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec2 with 772 individuals</span></span>
<span><span class="co">## tot_catch_vec: 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088 3088...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.67 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.05 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec1 with 743 individuals</span></span>
<span><span class="co">## tot_catch_vec: 743 743 743 743 743 743 743 743 743 743 743 743...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.64 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.05 seconds</span></span>
<span><span class="co">## Working on mixture collection: rec3 with 741 individuals</span></span>
<span><span class="co">## tot_catch_vec: 3839 3198 2791 4069 2834 2951 2911 2258 4157 2172 3606 4042...</span></span>
<span><span class="co">##   calculating log-likelihoods of the mixture individuals.   time: 0.06 seconds</span></span>
<span><span class="co">##   performing 2000 total sweeps, 100 of which are burn-in and will not be used in computing averages in method "MCMC"   time: 0.64 seconds</span></span>
<span><span class="co">##   tidying output into a tibble.   time: 0.05 seconds</span></span></code></pre>
<p>Let us now look at the resulting numbers, paying careful attention to
“rec1”, which is particularly illuminating. Because the total catch is
the same as before (743), but 300 of the mixture sample are <em>not</em>
part of the catch, then there is more variation in the total catch than
before.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_sstc_var</span> <span class="op">&lt;-</span> <span class="va">catch_ests_with_var</span><span class="op">$</span><span class="va">stock_specific_total_catch_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_sstc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">SSTC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_pprc_var</span> <span class="op">&lt;-</span> <span class="va">catch_ests_with_var</span><span class="op">$</span><span class="va">posterior_predictive_remaining_catch_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_pprc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">PPRC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>   <span class="co"># column PPRC has the posterior_predictive_remaining_catch</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CC_ac_var</span> <span class="op">&lt;-</span> <span class="va">catch_ests_with_var</span><span class="op">$</span><span class="va">allocation_count_traces</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>repunit_ac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">CA</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>  <span class="co"># column CA has the allocation_count</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sweep</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="va">repunit</span> <span class="op">==</span> <span class="st">"CaliforniaCoast"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'mixture_collection', 'sweep'. You can</span></span>
<span><span class="co">## override using the `.groups` argument.</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># then join those to the total stock-specific catch and pivot them</span></span>
<span><span class="va">three_counts_var</span> <span class="op">&lt;-</span> <span class="va">CC_sstc_var</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">CC_pprc_var</span>, by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html" class="external-link">join_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">CC_ac_var</span>, by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html" class="external-link">join_by</a></span><span class="op">(</span><span class="va">mixture_collection</span>, <span class="va">sweep</span>, <span class="va">repunit</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">repunit_sstc</span><span class="op">:</span><span class="va">repunit_ac</span>, names_to <span class="op">=</span> <span class="st">"count_type"</span>, values_to <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">three_counts_var</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n</span>, fill <span class="op">=</span> <span class="va">count_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">count_type</span> <span class="op">~</span> <span class="va">mixture_collection</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="uncertainty-on-total-catch_files/figure-html/morevar-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>That all looks as it should.</p>
</div>
<div class="section level3">
<h3 id="using-this-with-the-fully-bayesian-model">Using this with the “Fully-Bayesian” model<a class="anchor" aria-label="anchor" href="#using-this-with-the-fully-bayesian-model"></a>
</h3>
<p>The function <code><a href="../reference/infer_mixture.html">infer_mixture()</a></code> allows for a model in which
the allele frequencies are themselves updated in the MCMC by including
the gene copies of fish from the mixture that are allocated to each
collection. This model is invoked using the <code>method = "BR"</code>
option to <code><a href="../reference/infer_mixture.html">infer_mixture()</a></code>. The “BR” stands for “Bayesian
Resampling” (of allele frequencies). All the stock-specific total catch
estimation procedures described above have also been implemented in this
“BR” method. We don’t show them natively in the vignette because it
takes considerably longer and likely would not pass CRAN checks as a
consequence. However, you can run all the above steps using the “BR”
model by simply setting the variable <code>rubi_method &lt;- "BR"</code>
in the first R code block below the “Examples” section header.</p>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="citations">Citations<a class="anchor" aria-label="anchor" href="#citations"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0" line-spacing="2">
<div id="ref-cabanellas2017estimating" class="csl-entry">
Cabanellas-Reboredo, M., M. Palmer, J. Alos, and B. Morales-Nin. 2017.
Estimating harvest and its uncertainty in heterogeneous recreational
fisheries. Fisheries Research 188:100–111.
</div>
<div id="ref-faunce2015can" class="csl-entry">
Faunce, C. H., J. Cahalan, J. Bonney, and R. Swanson. 2015. Can observer
sampling validate industry catch reports from trawl fisheries? Fisheries
Research 172:34–43.
</div>
<div id="ref-rosenbaum2024reliability" class="csl-entry">
Rosenbaum, S. W., S. A. May, K. R. Shedd, C. J. Cunningham, R. L.
Peterson, B. W. Elliot, and M. V. McPhee. 2024. Reliability of
trans-generational genetic mark–recapture (tGMR) for enumerating pacific
salmon. Evolutionary Applications 17:e13647.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric C. Anderson, Ben Moran.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
