<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate genotypes by gene copy, with missing data from chosen individuals — gprob_sim_gc_missing • rubias</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate genotypes by gene copy, with missing data from chosen individuals — gprob_sim_gc_missing"><meta name="description" content="Takes a list of parameters from a genetic dataset, and returns a genotype log-likelihood
matrix for individuals simulated by gene copy from the specified collections, with
genotypes masked by missing data patterns from reference individuals"><meta property="og:description" content="Takes a list of parameters from a genetic dataset, and returns a genotype log-likelihood
matrix for individuals simulated by gene copy from the specified collections, with
genotypes masked by missing data patterns from reference individuals"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rubias</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rubias-fully-bayesian.html">Using the Fully Bayesian Model in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-overview.html">An Overview of rubias Usage</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-underlying-data-structures.html">An Explanation of the Underlying Data Structures in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/uncertainty-on-total-catch.html">Summarizing Uncertainty on Stock-Specific Total Catch</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/eriqande/rubias/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulate genotypes by gene copy, with missing data from chosen individuals</h1>
      <small class="dont-index">Source: <a href="https://github.com/eriqande/rubias/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>gprob_sim_gc_missing.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Takes a list of parameters from a genetic dataset, and returns a genotype log-likelihood
matrix for individuals simulated by gene copy from the specified collections, with
genotypes masked by missing data patterns from reference individuals</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gprob_sim_gc_missing</span><span class="op">(</span><span class="va">par_list</span>, <span class="va">sim_colls</span>, <span class="va">sim_missing</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-par-list">par_list<a class="anchor" aria-label="anchor" href="#arg-par-list"></a></dt>
<dd><p>genetic data converted to the param_list format by <code>tcf2param_list</code></p></dd>


<dt id="arg-sim-colls">sim_colls<a class="anchor" aria-label="anchor" href="#arg-sim-colls"></a></dt>
<dd><p>a vector; element i specifies the collection from which to sample
the genotypes for individual i</p></dd>


<dt id="arg-sim-missing">sim_missing<a class="anchor" aria-label="anchor" href="#arg-sim-missing"></a></dt>
<dd><p>a vector; element i specifies the index for the individual in
params$I whose missing data should be copied for individual i</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>In simulation by gene copy, the genotype at a locus for any individual is the result
of two random draws from the allele count matrix of that locus. Draws within an individual
are performed without replacement, but allele counts are replaced between individuals.
If the data at a particular locus is missing for individual i in <code>sim_missing</code>,
this data will also be missing in simulated individual i for the
log-likelihood calculation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If one wanted to simulate the missing data patterns</span></span></span>
<span class="r-in"><span><span class="co"># of a troublesome mixture dataset, one would run tcf2param_list,</span></span></span>
<span class="r-in"><span><span class="co"># selecting samp_type = "mixture", then draw sim_miss from</span></span></span>
<span class="r-in"><span><span class="co"># the mixture individual genotype list</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make a fake mixture data set to demonstrate...</span></span></span>
<span class="r-in"><span><span class="va">drawn</span> <span class="op">&lt;-</span> <span class="fu"><a href="mixture_draw.html">mixture_draw</a></span><span class="op">(</span><span class="va">alewife</span>, rhos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>,N <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">drawn</span><span class="op">$</span><span class="va">reference</span></span></span>
<span class="r-in"><span><span class="va">mix</span> <span class="op">&lt;-</span> <span class="va">drawn</span><span class="op">$</span><span class="va">mix</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># then run it...</span></span></span>
<span class="r-in"><span><span class="co"># we have to get the ploidies to pass to tcf2param_list</span></span></span>
<span class="r-in"><span><span class="va">locnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alewife</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">ploidies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">locnames</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ploidies</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">locnames</span></span></span>
<span class="r-in"><span><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="tcf2param_list.html">tcf2param_list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">ref</span>,<span class="va">mix</span><span class="op">)</span>, <span class="fl">17</span>, samp_type <span class="op">=</span> <span class="st">"mixture"</span>, ploidies <span class="op">=</span> <span class="va">ploidies</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Summary Statistics:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 99 Individuals in Sample</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 Loci: Aa046, Aa070, Aa074, Aa081, Aa091, Aa093, Ap010, Ap033, Ap038, Ap058, Ap071</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Reporting Units: NNE, SNE, MAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21 Collections: EMA, STG, PIS, MYS, MON, TBR, GIL, THA, BRI, CON, QUI, HOU, PEQ, MIA, HUD, DEL, NAN, RAP, CHO, ROA, ALL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5.14% of allelic data identified as missing</span>
<span class="r-in"><span><span class="va">sim_colls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">C</span>, <span class="fl">1070</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sim_miss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">coll</span><span class="op">)</span>, <span class="fl">1070</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ale_sim_gprobs_miss</span> <span class="op">&lt;-</span> <span class="fu">gprob_sim_gc_missing</span><span class="op">(</span><span class="va">params</span>, <span class="va">sim_colls</span>, <span class="va">sim_miss</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric C. Anderson, Ben Moran.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

