<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Collect essential data values before mixture proportion estimation — list_diploid_params • rubias</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Collect essential data values before mixture proportion estimation — list_diploid_params"><meta name="description" content="Takes all relevant information created in previous steps of data conversion pipeline,
and combines into a single list which serves as input for further calculations"><meta property="og:description" content="Takes all relevant information created in previous steps of data conversion pipeline,
and combines into a single list which serves as input for further calculations"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rubias</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/rubias-fully-bayesian.html">Using the Fully Bayesian Model in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-overview.html">An Overview of rubias Usage</a></li>
    <li><a class="dropdown-item" href="../articles/rubias-underlying-data-structures.html">An Explanation of the Underlying Data Structures in rubias</a></li>
    <li><a class="dropdown-item" href="../articles/uncertainty-on-total-catch.html">Summarizing Uncertainty on Stock-Specific Total Catch</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/eriqande/rubias/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Collect essential data values before mixture proportion estimation</h1>
      <small class="dont-index">Source: <a href="https://github.com/eriqande/rubias/blob/HEAD/R/data_conversion.R" class="external-link"><code>R/data_conversion.R</code></a></small>
      <div class="d-none name"><code>list_diploid_params.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Takes all relevant information created in previous steps of data conversion pipeline,
and combines into a single list which serves as input for further calculations</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">list_diploid_params</span><span class="op">(</span></span>
<span>  <span class="va">AC_list</span>,</span>
<span>  <span class="va">I_list</span>,</span>
<span>  <span class="va">PO</span>,</span>
<span>  <span class="va">coll_N</span>,</span>
<span>  <span class="va">RU_vec</span>,</span>
<span>  <span class="va">RU_starts</span>,</span>
<span>  alle_freq_prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>const_scaled <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ac-list">AC_list<a class="anchor" aria-label="anchor" href="#arg-ac-list"></a></dt>
<dd><p>a list of allele count matrices; output from <code>a_freq_list</code></p></dd>


<dt id="arg-i-list">I_list<a class="anchor" aria-label="anchor" href="#arg-i-list"></a></dt>
<dd><p>a list of genotype vectors; output from <code>allelic_list</code></p></dd>


<dt id="arg-po">PO<a class="anchor" aria-label="anchor" href="#arg-po"></a></dt>
<dd><p>a vector of collection (population of origin) indices
for every individual in the sample, in order identical to <code>I_list</code></p></dd>


<dt id="arg-coll-n">coll_N<a class="anchor" aria-label="anchor" href="#arg-coll-n"></a></dt>
<dd><p>a vector of the total number of individuals in each collection,
in order of appearance in the dataset</p></dd>


<dt id="arg-ru-vec">RU_vec<a class="anchor" aria-label="anchor" href="#arg-ru-vec"></a></dt>
<dd><p>a vector of collection indices, sorted by reporting unit</p></dd>


<dt id="arg-ru-starts">RU_starts<a class="anchor" aria-label="anchor" href="#arg-ru-starts"></a></dt>
<dd><p>a vector of indices, designating the first collection for each
reporting unit in RU_vec</p></dd>


<dt id="arg-alle-freq-prior">alle_freq_prior<a class="anchor" aria-label="anchor" href="#arg-alle-freq-prior"></a></dt>
<dd><p>a one-element named list specifying the prior to be used when
generating Dirichlet parameters for genotype likelihood calculations. The name of the
list item determines the type of prior used, with options <code>"const"</code>, <code>"scaled_const"</code>,
and <code>"empirical"</code>. If <code>"const"</code>, the listed number will be taken as a constant
added to the count for each allele, locus, and collection.
If <code>"scaled_const"</code>, the listed number will be divided by the number of alleles at a locus,
then added to the allele counts. If <code>"empirical"</code>, the listed number will be multiplied
by the relative frequency of each allele across all populations, then added to the allele counts.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>list_diploid_params</code> returns a list of the information necessary
for the calculation of genotype likelihoods in MCMC:</p>
<p><code>L</code>, <code>N</code>, and <code>C</code> represent the number of loci, individual genotypes,
and collections, respectively. <code>A</code> is a vector of the number of alleles at each
locus, and <code>CA</code> is the cumulative sum of <code>A</code>. <code>coll</code>, <code>coll_N</code>,
<code>RU_vec</code>, and <code>RU_starts</code> are copied directly from input.</p>
<p><code>I</code>, <code>AC</code>, <code>sum_AC</code>, <code>DP</code>, and <code>sum_DP</code> are vectorized
versions of data previously represented as lists and matrices; indexing macros
use <code>L</code>, <code>N</code>, <code>C</code>, <code>A</code>, and <code>CA</code> to access these vectors
in later Rcpp-based calculations.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Genotypes represented in <code>I_list</code> are converted into a single long vector,
ordered by locus, individual, and gene copy, with <code>NA</code> values represented as 0s.
Similarly, <code>AC_list</code> is unlisted to <code>AC</code>, ordered by locus, collection,
and allele. <code>DP</code> is a list of Dirichlet priors for likelihood calculations, created
by adding the values calculated from <code>alle_freq_prior</code> to each allele
<code>sum_AC</code> and <code>sum_DP</code> are the summed allele values for each locus
of their parent vectors, ordered by locus and collection.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">allelic_list</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> alllc_&gt; example(a_freq_list)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> a_frq_&gt;  # Generate a list of individual genotypes by allele from</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> a_frq_&gt;  # the alewife data's reference allele count tables</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> a_frq_&gt;  example(reference_allele_counts)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rfrn__&gt; ## count alleles in alewife reference populations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rfrn__&gt; example(tcf2long)  # gets variable ale_long</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; ## Convert the alewife dataset for further processing</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; # the data frame passed into this function must have had</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; # character collections and repunits converted to factors</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; reference &lt;- alewife</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; reference$repunit &lt;- factor(reference$repunit, levels = unique(reference$repunit))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; reference$collection &lt;- factor(reference$collection, levels = unique(reference$collection))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tcf2ln&gt; ale_long &lt;- tcf2long(reference, 17)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rfrn__&gt; ale_rac &lt;- reference_allele_counts(ale_long$long)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> a_frq_&gt;  ale_ac &lt;- a_freq_list(ale_rac)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> alllc_&gt; ale_cs &lt;- ale_long$clean_short</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> alllc_&gt; # Get the vectors of gene copies a and b for all loci in integer index form</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> alllc_&gt; ale_alle_list &lt;- allelic_list(ale_cs, ale_ac)$int</span>
<span class="r-in"><span><span class="va">PO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">ale_long</span><span class="op">$</span><span class="va">clean_short</span><span class="op">$</span><span class="va">collection</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coll_N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">PO</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">Colls_by_RU</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">ale_long</span><span class="op">$</span><span class="va">clean_short</span>, <span class="va">repunit</span>, <span class="va">collection</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="va">PC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">(</span><span class="va">Colls_by_RU</span><span class="op">$</span><span class="va">repunit</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Colls_by_RU</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>   <span class="va">PC</span><span class="op">[</span><span class="va">Colls_by_RU</span><span class="op">$</span><span class="va">repunit</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">PC</span><span class="op">[</span><span class="va">Colls_by_RU</span><span class="op">$</span><span class="va">repunit</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span></span></span>
<span class="r-in"><span> <span class="op">}</span></span></span>
<span class="r-in"><span><span class="va">RU_starts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">PC</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">RU_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">Colls_by_RU</span><span class="op">$</span><span class="va">collection</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">param_list</span> <span class="op">&lt;-</span> <span class="fu">list_diploid_params</span><span class="op">(</span><span class="va">ale_ac</span>, <span class="va">ale_alle_list</span>, <span class="va">PO</span>, <span class="va">coll_N</span>, <span class="va">RU_vec</span>, <span class="va">RU_starts</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric C. Anderson, Ben Moran.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

